{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Local Python iRIGS\n",
    "#### Author: Colleen Xu\n",
    "Winter - Spring 2020\n",
    "\n",
    "## Intro\n",
    "\n",
    "This is a Python3 Jupyter-notebook implementation of iRIGS (Integrative framework for Risk Gene Selector). **The notebook explains how the algorithms are implemented in detail, and it can be used to start modifying iRIGS for other studies/use cases/input data.** The [original R code](https://www.vumc.org/cgg/irigs) was released by the Computational Genetics and Genomics Lab at Vanderbilt University as part of the [*Nature Methods* paper](https://www.ncbi.nlm.nih.gov/pubmed/30988527):  \n",
    "\n",
    "Wang Q, Chen R, Cheng F, Wei Q, Ji Y, Yang H, et al. A Bayesian framework that integrates multi-omics data and gene networks predicts risk genes from schizophrenia GWAS data. Nature Neuroscience. 2019 May;22(5):691–9.  \n",
    "\n",
    "I relied heavily on the methods of this paper and its [Supplementary Note](https://www.nature.com/articles/s41593-019-0382-7#Sec26). **Below, I assume that the user of this notebook has downloaded the data from https://www.vumc.org/cgg/irigs and the notebook is in the same level/directory as the supporting_files directory and SNP_file directory.**      \n",
    "\n",
    "**My implementation differs from the original code in several ways. All major changes are noted and explained in detail throughout the notebook. A brief description of changes:**   \n",
    "* Different language (Python vs R)\n",
    "* Preprocessing the input gene information differently and before picking genes from flanking regions around GWAS SNPs   \n",
    "* Treating missing values for numbers of enhancers the same (replaced with zeroes)  \n",
    "* Gibbs sampling is not done on SNP loci with only one candidate gene.  \n",
    "* Reproducing the *Nature Methods* paper's Supplementary Table 1 as output  \n",
    "\n",
    "**Hanging Questions (not yet addressed):**  \n",
    "* There is a matrix included in the supporting files (go_propogation_probality_rp_0.3.RData). Is this matrix the transition probabilities, calculated from running random walk with restart until the stopping threshold was reached? Or is this the normalized weighted adjacency matrix for the GO-annotation-edge-weight network? \n",
    "* A normal distribution is assumed for the transformed variables created by the Mahalanobis transformation. However, my qq plots and normality tests on the variables fail to show that the variables are normally distributed. Would one proceed with using the normal distribution to get p-values (after all, the transformation is supposed to create variables that follow a normal distribution)? Or would one look to see if another distribution is a better fit?\n",
    "* The probability of a gene having a de novo mutation associated with schizophrenia (used as a p-value) is calculated  using a hard-coded denominator of 20,000. How was this number chosen (is it the assumed number of genes in the genome)?\n",
    "* What is the rationale for getting the negative log of the combined p-values (from Fisher's method) (extra_evi.R line 187)? I assumed that it was because the negative log would be a continuous measure of information, where a greater number would indicate more evidence that the candidate gene was a risk gene. (ref: https://lesslikely.com/statistics/s-values/)\n",
    "* I am confused by extra_evi.R line 202, where it appears that only one surrogate Bayes factor is kept per gene name. What is the rationale for this? My thinking is that since the surrogate Bayes factor includes \"distance to SNP\" as input, the factor differs depending on which SNP a given gene is associated with. I have therefore kept all the calculated surrogate Bayes factors in my implementation of iRIGS; during Gibbs sampling, the Bayes factors specific to the sampled SNP locus and candidate genes are pulled.\n",
    "\n",
    "### Description of iRIGS\n",
    "\n",
    "This framework was designed to find the genes \"most likely\" to be disease risk genes. Starting with schizophrenia (SCZ) GWAS data, the genes neighboring the SNPs (loci) are found. During Gibbs sampling, the candidate genes for a SNP loci are weighed based on:  \n",
    "\n",
    "* their \"closeness\" to the other sampled genes (around other SNP loci) in a gene-gene network made by the authors   \n",
    "  - with the code, the iRIGS authors provided what I believe is the transition probability matrix obtained by running the random walk with restart (RWR) algorithm on their original network. Their original network was dense, with edge weights between genes based on shared GO annotations. {? See the \"Hanging questions\" section above: Is this the transition probabilities from the RWR or just the normalized weighted adjacency matrix for the original network? We are missing code to generate this matrix and explore the original network.}  \n",
    "* their genomic distance from the SNP (distance from gene transcription start site (TSS) to SNP in genomic coordinates): abbreviated DTS (distance from transcription start)  \n",
    "  - based on gene information file provided by iRIGS authors {? We are missing metadata on where it was downloaded from and when}   \n",
    "* the number of enhancers (predicted distal regulatory element (DRE) - promoter links) they have  \n",
    "  - Brain HiC data from [the developing cerebral cortex - cortical/subcortical plate (CP) and germinal zone (GZ), Supplementary Tables 23 and 24](https://www.nature.com/articles/nature19847#Sec31): tsv versions provided by iRIGs authors     \n",
    "  - CapHiC data from the cell line GM12878 (not specific to brain): processed data file provided by iRIGs authors {? We are missing code on how to generate this file from \"raw data files\" from ArrayExpress}           \n",
    "  - FANTOM5 data: data file provided by iRIGs authors {? We are missing metadata on where it was downloaded from and when. I'm not sure if this includes all tissues or not.}   \n",
    "* de novo mutation (DNM) data for schizophrenia from literature: processed data files provided by iRIGS authors {? We are missing code on how to generate this file}   \n",
    "* differential expression (DE) data comparing schizophrenia vs control cases from the Commonmind consortium: data file provided by iRIGS authors {? It is unclear if this file was made manually or with a code pipeline}       \n",
    "\n",
    "**Assumptions:**  \n",
    "* It is assumed that a gene with more \"incoming regulatory links\" (enhancers) is more likely to be a disease gene (discussed in the original paper).  \n",
    "* All genomic coordinates are in hg19 (GRCh37).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages\n",
    "\n",
    "This code uses numpy, pandas, and scipy. It is designed to run simply with as few dependencies as possible. Some packages/tools are included for analyzing performance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats  ## use for pvalues, chidist\n",
    "\n",
    "from pathlib import Path  ## for dealing with path names regardless of machine (Windows/Mac/Linux)\n",
    "import math \n",
    "## get from https://anaconda.org/conda-forge/pyreadr\n",
    "import pyreadr  ## import RData objects! \n",
    "\n",
    "import re  ## for regular expressions\n",
    "import gc ## to delete/clean up dataframes I'm not using anymore\n",
    "import time  ## to time long code chunks \n",
    "\n",
    "## for printing every line that will give output\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## don't run unless you want to check your environment\n",
    "# !conda info --envs  ## check what environment I am in\n",
    "# !conda list ## check what packages I have\n",
    "\n",
    "## checking if line profiler is working: https://github.com/rkern/line_profiler\n",
    "# %lprun?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes on using pandas\n",
    "\n",
    "#### Using square brackets [ ] (the indexing operator) with pandas DataFrames:\n",
    "The behavior will be different depending on the object you put in the square brackets. [Reference](https://medium.com/dunder-data/selecting-subsets-of-data-in-pandas-39e811c81a0c)\n",
    "\n",
    "* string: return a column as a Series\n",
    "* list of strings: return all the columns named, in order, as a DataFrame\n",
    "* a slice of index labels or integers: select rows\n",
    "* **a sequence of booleans: select all rows where True.** Can be a list, tuple, pandas Series, numpy array. \n",
    "    - For a pandas Series, the index exactly match the DataFrame's index. \n",
    "    - For numpy arrays, it must be the same length as the DataFrame. \n",
    "\n",
    "I will be using this last option (a sequence of booleans) often. Because of the constraints, these sequences are almost always made from the data (DataFrame) by using a logical expression on the data. \n",
    "\n",
    "#### Using multiple conditions in Boolean indexing\n",
    "* Pandas will not accept the Python keywords (and, or, not). It will accept & (and), | (or), and ~ (not). \n",
    "* The multiple conditions can be written in one line inside the [ ]...but every condition **must be in parentheses** \n",
    "* consider breaking complicated indexing into multiple steps, saving the boolean sequences as variables\n",
    "\n",
    "#### Building a dataframe with a for loop (bind rows, append, concat):\n",
    "* It is more efficient to build a list of rows as iterables (tuples/dictionaries/lists), then turn this list into a dataframe. [Reference](https://stackoverflow.com/a/56746204)\n",
    "* Try not to iterate over a DataFrame when you can. [Reference](https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find candidate genes for each SNP \n",
    "\n",
    "**CHANGES from original code:** \n",
    "* I am filtering the human gene list BEFORE finding candidate genes for SNPs; this should make the search for candidate genes more efficient.  \n",
    "* I am filtering differently: original code relies heavily on the \"official_name\" column, but I will be working with ENSG IDs (\"Name\" column) and the \"alias_symbol\" column instead. There is an EDA code chunk showing the reasons for the change.   \n",
    "* I chose to drop certain genes because they were completely unconnected in the network. With the current framework design, these genes end up with a \"network\" factor of 0 and would not be chosen during the random sampling (unless all candidate genes for a SNP were in this situation - then a uniform probability of picking a gene would be implemented).  \n",
    "**Reminder: I assume that the user of this notebook has downloaded the data from https://www.vumc.org/cgg/irigs and the notebook is in the same level/directory as the supporting_files directory and SNP_file directory.**    \n",
    "\n",
    "### Load gene info, gene-gene network transition probabilities "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53934, 13)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(19080, 19080)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info_path = Path.cwd().joinpath('supporting_files', 'All_human_genes')\n",
    "gene_info = pd.read_csv(gene_info_path, sep = '\\t', header = 0)  \n",
    "gene_info.shape  ## 53934 rows, 13 columns\n",
    "\n",
    "## takes around 30 seconds to load gene-gene network matrix\n",
    "network_matrix_path = Path.cwd().joinpath('supporting_files', 'go_propogation_probality_rp_0.3.RData')\n",
    "result = pyreadr.read_r(str(network_matrix_path)) \n",
    "network_matrix = result['pro_p']\n",
    "\n",
    "network_genes = network_matrix.columns\n",
    "network_matrix.index = network_genes\n",
    "\n",
    "network_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up unnecessay large variable\n",
    "del result\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDA showing concerns with gene info's official_name column\n",
    "\n",
    "**Optional code chunks (don't have to run).** \n",
    "\n",
    "First, rows may have the same official_name values but different information (Name, alias_symbol, genomic coordinates). All rows have unique Name column values (Ensembl IDs), but there are some rows with shared official_name values and some rows with shared alias_symbol values.  \n",
    "\n",
    "I checked rows where official_name is \"AGAP4\" and where it was \"CX3CR1\". I found that **rows with the same official_name actually describe different genes. The IDs in the alias_symbol and Name columns are correct and reliable, and the IDs in the official_name column are not reliable.**  \n",
    "\n",
    "This can be explored in the code chunks below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Name values: 53934.\n",
      "Unique official_name values: 25752.\n",
      "Unique alias_symbol values: 49915.\n"
     ]
    }
   ],
   "source": [
    "## Look at number of unique identifers in columns with IDs\n",
    "print('Unique Name values: {}.'.format(len(pd.unique(gene_info.Name))))\n",
    "print('Unique official_name values: {}.'.format(len(pd.unique(gene_info.official_name))))\n",
    "print('Unique alias_symbol values: {}.'.format(len(pd.unique(gene_info.alias_symbol))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>official_name</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17511</th>\n",
       "      <td>ENSG00000188234.8</td>\n",
       "      <td>AGAP4</td>\n",
       "      <td>AGAP4</td>\n",
       "      <td>chr10</td>\n",
       "      <td>46321042</td>\n",
       "      <td>46349323</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17603</th>\n",
       "      <td>ENSG00000174194.9</td>\n",
       "      <td>AGAP4</td>\n",
       "      <td>AGAP8</td>\n",
       "      <td>chr10</td>\n",
       "      <td>51224677</td>\n",
       "      <td>51252947</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34240</th>\n",
       "      <td>ENSGR0000197976.6</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1660486</td>\n",
       "      <td>1671407</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32516</th>\n",
       "      <td>ENSG00000197976.6</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1710486</td>\n",
       "      <td>1721407</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17545</th>\n",
       "      <td>ENSG00000186807.9</td>\n",
       "      <td>ANXA8L1</td>\n",
       "      <td>ANXA8L2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>47746936</td>\n",
       "      <td>47763041</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25492</th>\n",
       "      <td>ENSG00000258611.1</td>\n",
       "      <td>YBX2</td>\n",
       "      <td>YBX2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>93608818</td>\n",
       "      <td>93609835</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32519</th>\n",
       "      <td>ENSG00000214717.4</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2404455</td>\n",
       "      <td>2419008</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34243</th>\n",
       "      <td>ENSGR0000214717.4</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2354455</td>\n",
       "      <td>2369008</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30399</th>\n",
       "      <td>ENSG00000179909.9</td>\n",
       "      <td>ZNF154</td>\n",
       "      <td>ZNF154</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58208735</td>\n",
       "      <td>58220579</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30398</th>\n",
       "      <td>ENSG00000249233.2</td>\n",
       "      <td>ZNF154</td>\n",
       "      <td>ZNF154</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58208735</td>\n",
       "      <td>58220562</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Name official_name alias_symbol  chrom  start_hg19  \\\n",
       "17511  ENSG00000188234.8         AGAP4        AGAP4  chr10    46321042   \n",
       "17603  ENSG00000174194.9         AGAP4        AGAP8  chr10    51224677   \n",
       "34240  ENSGR0000197976.6       AKAP17A      AKAP17A   chrY     1660486   \n",
       "32516  ENSG00000197976.6       AKAP17A      AKAP17A   chrX     1710486   \n",
       "17545  ENSG00000186807.9       ANXA8L1      ANXA8L2  chr10    47746936   \n",
       "...                  ...           ...          ...    ...         ...   \n",
       "25492  ENSG00000258611.1          YBX2         YBX2  chr15    93608818   \n",
       "32519  ENSG00000214717.4         ZBED1        ZBED1   chrX     2404455   \n",
       "34243  ENSGR0000214717.4         ZBED1        ZBED1   chrY     2354455   \n",
       "30399  ENSG00000179909.9        ZNF154       ZNF154  chr19    58208735   \n",
       "30398  ENSG00000249233.2        ZNF154       ZNF154  chr19    58208735   \n",
       "\n",
       "       end_hg19 strand  \n",
       "17511  46349323      -  \n",
       "17603  51252947      -  \n",
       "34240   1671407      +  \n",
       "32516   1721407      +  \n",
       "17545  47763041      +  \n",
       "...         ...    ...  \n",
       "25492  93609835      -  \n",
       "32519   2419008      -  \n",
       "34243   2369008      -  \n",
       "30399  58220579      -  \n",
       "30398  58220562      -  \n",
       "\n",
       "[140 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## find rows of gene information with the same official_name values\n",
    "ex0 = gene_info[gene_info.duplicated(subset = 'official_name', keep = False)]  \n",
    "## the duplicated method returns a boolean vector of whether every row contains a duplicated value\n",
    "## arguments: subset tells what columns to look for duplicate values in. \n",
    "##            keep = False parameter will mark all duplicates. \n",
    "\n",
    "## restrict view to useful columns: see how other columns' values are diff for same official_name! \n",
    "ex0 = ex0.filter(items = ['Name', 'official_name', 'alias_symbol', 'chrom', 'start_hg19', 'end_hg19', 'strand'])\n",
    "\n",
    "## restrict view to rows where official_name of the gene is in network\n",
    "## then arrange rows by the official_name values\n",
    "ex0[ex0['official_name'].isin(network_genes)].sort_values(by = 'official_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Second, if I remove rows where the official_name value is missing (NaN), I lose information on genes in my network.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>official_name</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44681</th>\n",
       "      <td>ENSG00000158122.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAED1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>99401859</td>\n",
       "      <td>99417599</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51555</th>\n",
       "      <td>ENSG00000161533.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACOX1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>73937596</td>\n",
       "      <td>73975515</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45554</th>\n",
       "      <td>ENSG00000151303.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88728247</td>\n",
       "      <td>88784489</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40596</th>\n",
       "      <td>ENSG00000189045.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANKDD1B</td>\n",
       "      <td>chr5</td>\n",
       "      <td>74907301</td>\n",
       "      <td>74967671</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42321</th>\n",
       "      <td>ENSG00000157999.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANKRD61</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6071007</td>\n",
       "      <td>6076017</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51892</th>\n",
       "      <td>ENSG00000213297.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF625</td>\n",
       "      <td>chr19</td>\n",
       "      <td>12242932</td>\n",
       "      <td>12267546</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43258</th>\n",
       "      <td>ENSG00000215356.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF705B</td>\n",
       "      <td>chr8</td>\n",
       "      <td>7783859</td>\n",
       "      <td>7812271</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42658</th>\n",
       "      <td>ENSG00000234444.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF736</td>\n",
       "      <td>chr7</td>\n",
       "      <td>63767837</td>\n",
       "      <td>63810017</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52087</th>\n",
       "      <td>ENSG00000261221.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF865</td>\n",
       "      <td>chr19</td>\n",
       "      <td>56116771</td>\n",
       "      <td>56128635</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44769</th>\n",
       "      <td>ENSG00000228623.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF883</td>\n",
       "      <td>chr9</td>\n",
       "      <td>115759495</td>\n",
       "      <td>115774507</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name official_name alias_symbol  chrom  start_hg19  \\\n",
       "44681   ENSG00000158122.6           NaN        AAED1   chr9    99401859   \n",
       "51555   ENSG00000161533.5           NaN        ACOX1  chr17    73937596   \n",
       "45554  ENSG00000151303.10           NaN       AGAP11  chr10    88728247   \n",
       "40596   ENSG00000189045.7           NaN      ANKDD1B   chr5    74907301   \n",
       "42321   ENSG00000157999.5           NaN      ANKRD61   chr7     6071007   \n",
       "...                   ...           ...          ...    ...         ...   \n",
       "51892   ENSG00000213297.4           NaN       ZNF625  chr19    12242932   \n",
       "43258   ENSG00000215356.3           NaN      ZNF705B   chr8     7783859   \n",
       "42658   ENSG00000234444.3           NaN       ZNF736   chr7    63767837   \n",
       "52087   ENSG00000261221.1           NaN       ZNF865  chr19    56116771   \n",
       "44769   ENSG00000228623.2           NaN       ZNF883   chr9   115759495   \n",
       "\n",
       "        end_hg19 strand  \n",
       "44681   99417599      -  \n",
       "51555   73975515      -  \n",
       "45554   88784489      +  \n",
       "40596   74967671      +  \n",
       "42321    6076017      +  \n",
       "...          ...    ...  \n",
       "51892   12267546      -  \n",
       "43258    7812271      +  \n",
       "42658   63810017      +  \n",
       "52087   56128635      +  \n",
       "44769  115774507      -  \n",
       "\n",
       "[140 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## restrict view to rows where...\n",
    "ex0[ex0['alias_symbol'].isin(network_genes) &    ## alias_symbol of the gene is in network\n",
    "    ex0['official_name'].isna()                     ## official_name of gene is missing\n",
    "   ].sort_values('alias_symbol')  ## arrange rows by alias_symbol column values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up the dataframe\n",
    "del ex0\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Must Run: EDA showing unconnected genes in network  \n",
    "\n",
    "Must run this chunk so the filtering in the next code chunk works. \n",
    "\n",
    "There are 1296 genes that have row and column sums of 0.3, with the self-edge (gene A - gene A entry in the matrix) being 0.3. This means that when Gibbs sampling is done, their network factor as a candidate gene is always 0 - leading to a probability weighting of 0 since self-edge weights are not included. The genes are therefore NEVER being picked in a sample. It is then inefficient to keep these genes in the calculation, so I chose to filter them out. \n",
    "\n",
    "The code chunk below can be used to explore this. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.300000    1296\n",
       "1.000114     211\n",
       "0.999857      77\n",
       "0.999857      73\n",
       "0.999857      58\n",
       "            ... \n",
       "0.999943       1\n",
       "0.999941       1\n",
       "0.999903       1\n",
       "0.999931       1\n",
       "0.999886       1\n",
       "Length: 15803, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "AAED1       0.3\n",
       "AAR2        0.3\n",
       "ABRACL      0.3\n",
       "ACTL10      0.3\n",
       "AKAP2       0.3\n",
       "           ... \n",
       "YPEL5       0.3\n",
       "ZBED6CL     0.3\n",
       "ZC3HAV1L    0.3\n",
       "ZYG11A      0.3\n",
       "ZYG11B      0.3\n",
       "Length: 1296, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "0.0    19079\n",
       "0.3        1\n",
       "Name: AAED1, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "0.3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "0.3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## some of the column sums of the matrix are only 0.3 (the probability that the walk stays at the current node), \n",
    "##        which is odd. The other columns sum ~ 1. \n",
    "## takes me ~10 seconds to run \n",
    "networkcolsums = network_matrix.sum(axis = 0)  ## apply to columns\n",
    "networkcolsums.value_counts()\n",
    "\n",
    "## some of the columns with this odd column sum\n",
    "oddcolsums = networkcolsums[networkcolsums == 0.3]\n",
    "oddcolsums\n",
    "\n",
    "## looking into column 'AAED1': \n",
    "network_matrix['AAED1'].value_counts()  ## there is only 1 non-zero value: 0.3\n",
    "network_matrix.loc['AAED1','AAED1']  ## 0.3 is the probability of staying in this node\n",
    "\n",
    "## look at the row sum for this gene: same issue - the values relating this node to other nodes is 0\n",
    "network_matrix.loc['AAED1',].sum()\n",
    "## I looked at the row sums for AAR2, ABRACL, ZYG11A: all are 0.3. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering to get the input gene set (to look for candidates within)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17784, 17784)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(17062, 13)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "17017"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## takes me a few seconds to run to drop the columns/rows\n",
    "\n",
    "## keep only the rows where the gene symbol ('alias_symbol column) is in my network \n",
    "unconnected = oddcolsums.index\n",
    "## get genes that are connected: removes 1296 genes to leave 17,784\n",
    "network_matrix.drop(unconnected, axis = 'index', inplace = True)\n",
    "network_matrix.drop(unconnected, axis = 'columns', inplace = True)\n",
    "# network_matrix.index = network_matrix.columns  ## in case rownames got removed...\n",
    "network_genes = network_matrix.columns  ## update\n",
    "network_matrix.shape\n",
    "\n",
    "## I want input genes to be in the network (so I have network information)\n",
    "gene_info = gene_info[gene_info['alias_symbol'].isin(network_genes)]\n",
    "\n",
    "gene_info.shape \n",
    "gene_info['alias_symbol'].nunique()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are now 17062 rows in gene_info with 17017 unique gene symbols (alias_symbol). The below code chunk shows that there are 90 rows / 2 rows per alias symbol = 45 unique alias symbols. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(90, 13)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "alias_symbol\n",
       "AKAP17A      2\n",
       "ASMT         2\n",
       "ASMTL        2\n",
       "C17orf49     2\n",
       "C20orf141    2\n",
       "CD99         2\n",
       "CLEC19A      2\n",
       "CLLU1        2\n",
       "CORO7        2\n",
       "CRLF2        2\n",
       "CSF2RA       2\n",
       "CYCS         2\n",
       "DHRSX        2\n",
       "DIO3         2\n",
       "DTX2         2\n",
       "GATC         2\n",
       "GTPBP6       2\n",
       "IL3RA        2\n",
       "IL9R         2\n",
       "KIAA1456     2\n",
       "LIMS3        2\n",
       "MFRP         2\n",
       "MRPL46       2\n",
       "MTG1         2\n",
       "NOL12        2\n",
       "OR11H12      2\n",
       "OR1D5        2\n",
       "OVCA2        2\n",
       "P2RY8        2\n",
       "PLCXD1       2\n",
       "PPP2R3B      2\n",
       "RNASE12      2\n",
       "RNASE4       2\n",
       "RTEL1        2\n",
       "SAMD4A       2\n",
       "SEC16B       2\n",
       "SHOX         2\n",
       "SLC25A6      2\n",
       "SPRY3        2\n",
       "SULT1A3      2\n",
       "VAMP7        2\n",
       "YBX2         2\n",
       "ZBED1        2\n",
       "ZNF154       2\n",
       "ZNF625       2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## what is going on with these rows with the same alias_symbol value? Looking...\n",
    "ex1 = gene_info[gene_info.duplicated(subset='alias_symbol', keep=False)].sort_values(by='alias_symbol')\n",
    "ex1.shape\n",
    "ex1.groupby('alias_symbol').size() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After [looking up](http://grch37.ensembl.org/Homo_sapiens/Info/Index) through the Name column (ENSEMBL IDs), I found that some 'duplicate' alias symbols were different transcripts - where one was matched the Ensembl version. I went through those manually to create the array of IDs that matched the Ensembl version. \n",
    "\n",
    "The rest of the 'duplicate' alias symbols are legit. They are pseudoautosomal regions, which are matched regions on the X and Y chromosomes. These genes therefore have two loci/sets of genomic coordinates for the SAME gene symbol. Source (has more genes than those in my list below): https://www.genenames.org/data/genegroup/#!/group/714"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17035, 13)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "17017"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## IDs that match the Ensembl version\n",
    "correct_versions = np.array([\n",
    "    'ENSG00000258315.1',  # C17orf49\n",
    "    'ENSG00000258713.1',  # C20orf141: this row in df has different coordinates from ENSEMBL (ENSEMBL has Chromosome 20: 2,795,633-2,796,479) \n",
    "    'ENSG00000261210.1',  # CLEC19A\n",
    "    'ENSG00000257127.1',  # CLLU1\n",
    "    'ENSG00000262246.1',  # CORO7\n",
    "    'ENSG00000172115.4',  # CYCS\n",
    "    'ENSG00000197406.5',  # DIO3\n",
    "    'ENSG00000091073.13', # DTX2\n",
    "    'ENSG00000257218.1',  # GATC\n",
    "    'ENSG00000250305.3',  # KIAA1456\n",
    "    'ENSG00000256977.2',  # LIMS3: this row in df has different coordinates from ENSEMBL (ENSEMBL has Chromosome 2: 110,656,005-110,677,180) \n",
    "    'ENSG00000235718.3',  # MFRP: this alias has two legit-looking ENSG IDs, I picked the one with more transcripts/orthologs\n",
    "    'ENSG00000259494.1',  # MRPL46\n",
    "    'ENSG00000148824.12', # MTG1\n",
    "    'ENSG00000100101.12', # NOL12: this row in df has different coordinates from ENSEMBL (ENSEMBL has Chromosome 22: 38,077,680-38,170,137) \n",
    "    'ENSG00000257115.1',  # OR11H12\n",
    "    'ENSG00000262628.1',  # OR1D5\n",
    "    'ENSG00000262664.1',  # OVCA2\n",
    "    'ENSG00000258436.1',  # RNASE12\n",
    "    'ENSG00000258818.1',  # RNASE4\n",
    "    'ENSG00000258366.2',  # RTEL1\n",
    "    'ENSG00000020577.9',  # SAMD4A\n",
    "    'ENSG00000120341.10', # SEC16B: this row in df has different coordinates from ENSEMBL (ENSEMBL has Chromosome 1: 177,893,091-177,953,438) \n",
    "    'ENSG00000261052.1',  # SULT1A3 \n",
    "    'ENSG00000006047.8',  # YBX2\n",
    "    'ENSG00000179909.9',  # ZNF154\n",
    "    'ENSG00000257591.1'   # ZNF625\n",
    "], dtype=object)\n",
    "\n",
    "## Pseudoautosomal genes\n",
    "pseudoautosomal_list = np.array(['AKAP17A', 'ASMT', 'ASMTL', 'CD99', 'CRLF2', 'CSF2RA', 'DHRSX', 'GTPBP6',\n",
    "                                 'IL3RA', 'IL9R', 'P2RY8', 'PLCXD1', 'PPP2R3B', 'SHOX', 'SLC25A6', 'SPRY3', \n",
    "                                 'VAMP7', 'ZBED1'], dtype=object)  \n",
    "\n",
    "## for dataframe isin() method: \n",
    "## dict key is column name, value is list of values to check for in that column \n",
    "criteria = {'Name': correct_versions, 'alias_symbol': pseudoautosomal_list}\n",
    "## saves rows we don't want to keep: it's NOT the correct Ensembl ID OR it's NOT a pseudoautosomal gene\n",
    "ex1 = ex1[~ ex1.isin(criteria).any('columns')]  \n",
    "## reference: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html?highlight=isin#indexing-with-isin\n",
    "\n",
    "## get dataframe with rows from ex1 removed\n",
    "gene_info = gene_info[~ gene_info['Name'].isin(ex1['Name'])]\n",
    "\n",
    "## stats for filtered dataframe of gene info  \n",
    "gene_info.shape\n",
    "gene_info['alias_symbol'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final gene_info dataframe only has genes in the connected part of the gene-gene network. There are 17035 rows (unique Ensembl IDs in the 'Name' column) and 17017 unique gene symbols (alias_symbol column). The duplicate gene symbols are from 18 pseudoautosomal genes.    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up the dataframe\n",
    "del ex1, criteria, correct_versions, unconnected, networkcolsums, oddcolsums\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get GWAS SNP info\n",
    "**CHANGES from original code:** removed check on flank_length > 0  \n",
    "\n",
    "**Notes on parameters:**\n",
    "* SNP_file_path: path of file with GWAS SNPs. The file must include three columns named (case matters): SNP, Chr, and Pos_hg19  \n",
    "* flank: integer (in bp, recommended by iRIGS authors: 1000000 for a 2MB flanking region total). The flanking region is the genomic region between *flank* base-pairs upstream and *flank* base-pairs downstream of a GWAS SNP. In other words, *flank* is 1/2 the length of the flanking region. Genes in the flanking region will be used as candidate genes during Gibbs sampling    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Parameters\n",
    "SNP_file_path = Path.cwd().joinpath('SNP_file', 'SCZ_108_loci')\n",
    "flank = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Pos_hg19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1702294</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98501984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs11191419</td>\n",
       "      <td>chr10</td>\n",
       "      <td>104612335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2007044</td>\n",
       "      <td>chr12</td>\n",
       "      <td>2344960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4129585</td>\n",
       "      <td>chr8</td>\n",
       "      <td>143312933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP    Chr   Pos_hg19\n",
       "0  rs115329265   chr6   28712247\n",
       "1    rs1702294   chr1   98501984\n",
       "2   rs11191419  chr10  104612335\n",
       "3    rs2007044  chr12    2344960\n",
       "4    rs4129585   chr8  143312933"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(108, 3)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## There was a check here to halt the program if the variable flank < 0: I removed it\n",
    "\n",
    "## read GWAS SNP file: this will go forward w/o error even if the columns specified in dtype aren't there\n",
    "df_gwas_snps = pd.read_csv(SNP_file_path, sep='\\t', header=0, \\\n",
    "                           dtype={'SNP': object, 'Chr': object, 'Pos_hg19': 'Int64'})\n",
    "\n",
    "## check if the required columns are in the table\n",
    "if not {'SNP', 'Chr', 'Pos_hg19'}.issubset(df_gwas_snps.columns):\n",
    "    raise KeyError(\\\n",
    "        'At least one of the following required columns is missing from the GWAS SNP file: SNP, Chr, Pos_hg19')\n",
    "\n",
    "## if needed, in Chr column, add \"chr\" prefix to each chromosome name \n",
    "## grabs first entry of Chr column and checks that it starts with chr\n",
    "if not df_gwas_snps['Chr'][1].startswith('chr'):  \n",
    "    df_gwas_snps['Chr'] = 'chr' + df_gwas_snps['Chr']\n",
    "\n",
    "## look at dataframe to check that it was imported/processed correctly\n",
    "df_gwas_snps.head()\n",
    "df_gwas_snps.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find candidate genes around GWAS SNPs      \n",
    "\n",
    "Because pseudoautosomal genes are on different chromosomes, this method of getting candidate genes that neighbor SNPs will NEVER pick up both copies of a pseudoautosomal gene. **This means that within each candidate gene set for a SNP, alias_symbol values will be unique and we can use this column going forward to ID genes/rows.** \n",
    "\n",
    "Note that this version (vectorized, running on numpy arrays) runs about 5X faster than other methods (stored in the commented code chunk). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%timeit  ## for timing\n",
    "\n",
    "candidate_genes = pd.merge(df_gwas_snps, gene_info, left_on = 'Chr', right_on = 'chrom', how=\"left\")\n",
    "## creates dataframe where genes are appended to SNPs if chromosome number matches\n",
    "\n",
    "def snp_to_candidates(SNP_name, SNP_pos, gene_start, gene_end, flank_length):\n",
    "    \"\"\"\n",
    "    Arguments: \n",
    "    all numpy arrays must be the same length, ideally converted from columns of the same pandas DataFrame\n",
    "    \n",
    "    SNP_name: numpy array of strings, SNP names\n",
    "    SNP_pos: numpy array of int, SNP position in hg19\n",
    "    gene_start: numpy array of int, gene start position in hg19\n",
    "    gene_end: numpy array of int, gene end position in hg19\n",
    "    flank_length: int, from original arguments \n",
    "                  range on one side of SNP to look for candidate genes\n",
    "    \n",
    "    Returns: pandas Boolean array (same length as input arrays) for whether genes are within flanking region\n",
    "             and are therefore candidate genes \n",
    "    \"\"\"\n",
    "    ## to add/subtract flank from SNP position array, need this in array form\n",
    "    flank_vector = np.repeat(flank_length, len(SNP_pos))\n",
    "    \n",
    "    ## get arrays with flank start and end for SNP in each row\n",
    "    flank_start = np.maximum(SNP_pos - flank_vector, 0)  ## beginning of chromosome or upstream end\n",
    "    flank_end = SNP_pos + flank_vector \n",
    "    ## downstream end: doesn't cause error if this is beyond length of chromosome\n",
    "    \n",
    "    ## boolean vector: True if gene starts after flank_start and before flank_end.\n",
    "    criteria_start = (gene_start > flank_start) & (gene_start < flank_end)\n",
    "    ## boolean vector: True if gene ends after flank_start and before flank_end.\n",
    "    criteria_end = (gene_end > flank_start) & (gene_end < flank_end)\n",
    "    ## logical OR. This means that genes that partially overlap the flanking region \n",
    "    ## (end in flanking region or begin in flanking region) are also included\n",
    "    index_logic = criteria_start | criteria_end\n",
    "    return index_logic\n",
    "\n",
    "# # used to measure performance of function: notice the function only runs once\n",
    "# %lprun -f snp_to_candidates temptry_logic = snp_to_candidates(candidate_genes['SNP'].values, \\\n",
    "#                                                               candidate_genes['Pos_hg19'].values, \\\n",
    "#                                                               candidate_genes['start_hg19'].values,\\\n",
    "#                                                               candidate_genes['end_hg19'].values, flank)\n",
    "\n",
    "candidate_logic = snp_to_candidates(candidate_genes['SNP'].values, \\\n",
    "                                    candidate_genes['Pos_hg19'].values, \\\n",
    "                                    candidate_genes['start_hg19'].values,\\\n",
    "                                    candidate_genes['end_hg19'].values, \n",
    "                                    flank)\n",
    "\n",
    "candidate_genes = candidate_genes[candidate_logic]\n",
    "candidate_genes.rename(columns = {'Chr':'SNP_chr', 'Pos_hg19':'SNP_pos_hg19'}, inplace = True)\n",
    "## chunk timing\n",
    "# 85.6 ms ± 704 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is other attempts in \"vectorizing\". It gets the same results as the above code. However, it is slower. Uncomment to run and see timing. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%timeit  \n",
    "\n",
    "# def snp_to_candidates_slower(SNP_name, chr_num, pos_hg19, flank_length, gene_df):\n",
    "#     \"\"\"\n",
    "#     Arguments:\n",
    "#     SNP_name: string, name of the SNP\n",
    "#     chr_num: string, format: \"chr1\"\n",
    "#     pos_hg19: int, position in hg19 \n",
    "#     flank_length: int, from original arguments \n",
    "#                   range on one side of SNP to look for candidate genes\n",
    "#     gene_df: dataframe with gene info. MUST HAVE chrom (string, format \"chr1\"), \n",
    "#              start_hg19 (int), end_hg19 (int) columns (position info). \n",
    "#     \"\"\"\n",
    "#     flank_start = max(pos_hg19 - flank_length, 0)  ## beginning of chromosome or upstream end\n",
    "#     flank_end = pos_hg19 + flank_length \n",
    "#     ## downstream end: doesn't cause error if this is beyond length of chromosome\n",
    "    \n",
    "#     ## subset to find genes in same chromosome as SNP\n",
    "#     temp = gene_df[gene_df['chrom'] == chr_num]  \n",
    "#     ## boolean vector: True if gene starts after flank_start and before flank_end.\n",
    "#     criteria_start = (temp['start_hg19'] > flank_start) & (temp['start_hg19'] < flank_end)\n",
    "#     ## boolean vector: True if gene ends after flank_start and before flank_end.\n",
    "#     criteria_end = (temp['end_hg19'] > flank_start) & (temp['end_hg19'] < flank_end)\n",
    "#     ## logical OR. This means that genes that partially overlap the flanking region \n",
    "#     ## (end in flanking region or begin in flanking region) are also included\n",
    "#     temp = temp[criteria_start | criteria_end]\n",
    "    \n",
    "#     if len(temp) > 0:  ## if candidate genes were found\n",
    "#         ## append matched GWAS SNP info to each row (gene)\n",
    "#         temp['SNP'] = SNP_name\n",
    "#         temp['SNP_chr'] = chr_num\n",
    "#         temp['SNP_pos_hg19'] = pos_hg19\n",
    "        \n",
    "#         ## get a list of rows (each row is a list)\n",
    "#         temp_list = temp.values.tolist()  ## tolist is a numpy function\n",
    "#         return temp_list\n",
    "#     else:\n",
    "#         return []  ## return empty list\n",
    "\n",
    "# candidate_genes_list = []\n",
    "\n",
    "# ## apply function over rows\n",
    "# # templist = df_gwas_snps.apply(lambda row: snp_to_candidates_slower(row['SNP'], row['Chr'], \\\n",
    "# #                                                             row['Pos_hg19'], flank, df_gene_info), axis=1)\n",
    "# # for alist in templist:\n",
    "# #     candidate_genes_list.extend(alist)\n",
    "# ## speed: 496 ms ± 10.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "# ## iterate over rows: =\n",
    "# for row in df_gwas_snps.itertuples(index=False):\n",
    "#     temp = snp_to_candidates_slower(row.SNP, row.Chr, row.Pos_hg19, flank, df_gene_info)\n",
    "#     candidate_genes_list.extend(temp)\n",
    "# ## speed: 476 ms ± 7.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "# candidate_genes_slower = pd.DataFrame(candidate_genes_list, columns = ['Name', 'Description', 'chrom', 'type', \n",
    "#                                                                 'start_hg19', 'end_hg19','strand', 'ts_id', \n",
    "#                                                                 'gene_type', 'gene_status', 'loci_level',\n",
    "#                                                                 'alias_symbol', 'official_name', 'SNP', \n",
    "#                                                                'SNP_chr', 'SNP_pos_hg19'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1637, 16)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>ts_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130.4</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185130.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611.3</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182611.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374.5</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196374.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822.2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000233822.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357.3</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184357.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             SNP SNP_chr  SNP_pos_hg19               Name Description chrom  \\\n",
       "133  rs115329265    chr6      28712247  ENSG00000185130.4   HIST1H2BL  chr6   \n",
       "134  rs115329265    chr6      28712247  ENSG00000182611.3   HIST1H2AJ  chr6   \n",
       "135  rs115329265    chr6      28712247  ENSG00000196374.5   HIST1H2BM  chr6   \n",
       "136  rs115329265    chr6      28712247  ENSG00000233822.2   HIST1H2BN  chr6   \n",
       "137  rs115329265    chr6      28712247  ENSG00000184357.3    HIST1H1B  chr6   \n",
       "\n",
       "     type  start_hg19  end_hg19 strand              ts_id       gene_type  \\\n",
       "133  gene    27775257  27775709      -  ENSG00000185130.4  protein_coding   \n",
       "134  gene    27782112  27782607      -  ENSG00000182611.3  protein_coding   \n",
       "135  gene    27782822  27783267      +  ENSG00000196374.5  protein_coding   \n",
       "136  gene    27806323  27819974      +  ENSG00000233822.2  protein_coding   \n",
       "137  gene    27834570  27835359      -  ENSG00000184357.3  protein_coding   \n",
       "\n",
       "    gene_status  loci_level alias_symbol official_name  \n",
       "133       KNOWN           2    HIST1H2BL     HIST1H2BL  \n",
       "134       KNOWN           2    HIST1H2AJ     HIST1H2AJ  \n",
       "135       KNOWN           2    HIST1H2BM     HIST1H2BM  \n",
       "136       KNOWN           2    HIST1H2BN     HIST1H2BN  \n",
       "137       KNOWN           2     HIST1H1B      HIST1H1B  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1577 unique ENSG IDs (genes) near 107 GWAS SNPs were found...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## look at candidate gene dataframe\n",
    "candidate_genes.shape\n",
    "candidate_genes.head()\n",
    "\n",
    "## save numpy arrays with SNPs and genes (no duplicates)\n",
    "SNP_loci = candidate_genes['SNP'].unique()\n",
    "candidate_genelist = candidate_genes['alias_symbol'].unique()\n",
    "\n",
    "print('{genes} unique ENSG IDs (genes) near {loci} GWAS SNPs were found...\\n'.format(\n",
    "      genes = len(candidate_genelist), loci = len(SNP_loci)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 1637 rows in the candidate_genes dataframe but only 1577 unique genes, which means that **some SNP loci have the same candidate genes**. As shown in the code chunk below, there are no pseudoautosomal genes in this set. Also, one SNP had no candidate genes and is not in the dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EDA: \n",
    "## There are no pseudoautosomal genes in this set.   \n",
    "candidate_genes[candidate_genes['alias_symbol'].isin(pseudoautosomal_list)].empty\n",
    "## trying to get a subset of pseudoautosomal genes -> get an empty dataframe \n",
    "\n",
    "## one SNP had no candidate genes found: chr2_146436222_I\n",
    "candidate_genes.query(\"SNP == 'chr2_146436222_I'\").empty\n",
    "## trying to get a subset for this SNP -> get an empty dataframe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up \n",
    "del candidate_logic\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing network and supporting evidence\n",
    "\n",
    "### Filter network adjacency matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1577, 1577)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## slice the matrix to only include candidate genes\n",
    "network_matrix = network_matrix.loc[candidate_genelist, candidate_genelist]  ## end up (1577, 1577)\n",
    "network_matrix.shape\n",
    "\n",
    "## to match ENSG IDs in supporting evidence files, take version number off ENSG IDs \n",
    "candidate_genes['Name'] = candidate_genes['Name'].str.slice(0, 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 1: distance from SNP to gene transcription start site (TSS) (DTS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_DTS(SNP_pos, gene_strand, gene_start, gene_end):\n",
    "    \"\"\"\n",
    "    Arguments: \n",
    "    all arrays must be the same length, ideally columns from the same pandas DataFrame\n",
    "    \n",
    "    SNP_pos: numpy array of int, SNP position in hg19\n",
    "    gene_strand: numpy array of strings, what strand the gene is on\n",
    "    gene_start: numpy array of int, gene start position in hg19\n",
    "    gene_end: numpy array of int, gene end position in hg19\n",
    "    \n",
    "    Returns: an numpy array of int, linear sequence distance from SNP\n",
    "    \"\"\"\n",
    "    ## want to iterate over indices for these arrays\n",
    "    array_length = len(SNP_pos)\n",
    "\n",
    "    ## list to build \n",
    "    DTS = []\n",
    "    \n",
    "    for idx in range(array_length):\n",
    "        if gene_strand[idx] == '+':  # if gene is on strand + \n",
    "            temp = abs(gene_start[idx] - SNP_pos[idx])  # TSS is at gene start\n",
    "        else:\n",
    "            temp = abs(gene_end[idx] - SNP_pos[idx])  # TSS is at gene end\n",
    "        DTS.append(temp)\n",
    "    \n",
    "    DTS = np.array(DTS)\n",
    "    return DTS\n",
    "\n",
    "## used to measure performance of function\n",
    "candidate_genes['DTS'] = get_DTS(candidate_genes['SNP_pos_hg19'].values, \\\n",
    "                                          candidate_genes['strand'].values, \\\n",
    "                                          candidate_genes['start_hg19'].values,\\\n",
    "                                          candidate_genes['end_hg19'].values) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>ts_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "      <th>DTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185130.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>936538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182611.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>929640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196374.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>929425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000233822.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>905924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184357.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>876888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             SNP SNP_chr  SNP_pos_hg19             Name Description chrom  \\\n",
       "133  rs115329265    chr6      28712247  ENSG00000185130   HIST1H2BL  chr6   \n",
       "134  rs115329265    chr6      28712247  ENSG00000182611   HIST1H2AJ  chr6   \n",
       "135  rs115329265    chr6      28712247  ENSG00000196374   HIST1H2BM  chr6   \n",
       "136  rs115329265    chr6      28712247  ENSG00000233822   HIST1H2BN  chr6   \n",
       "137  rs115329265    chr6      28712247  ENSG00000184357    HIST1H1B  chr6   \n",
       "\n",
       "     type  start_hg19  end_hg19 strand              ts_id       gene_type  \\\n",
       "133  gene    27775257  27775709      -  ENSG00000185130.4  protein_coding   \n",
       "134  gene    27782112  27782607      -  ENSG00000182611.3  protein_coding   \n",
       "135  gene    27782822  27783267      +  ENSG00000196374.5  protein_coding   \n",
       "136  gene    27806323  27819974      +  ENSG00000233822.2  protein_coding   \n",
       "137  gene    27834570  27835359      -  ENSG00000184357.3  protein_coding   \n",
       "\n",
       "    gene_status  loci_level alias_symbol official_name     DTS  \n",
       "133       KNOWN           2    HIST1H2BL     HIST1H2BL  936538  \n",
       "134       KNOWN           2    HIST1H2AJ     HIST1H2AJ  929640  \n",
       "135       KNOWN           2    HIST1H2BM     HIST1H2BM  929425  \n",
       "136       KNOWN           2    HIST1H2BN     HIST1H2BN  905924  \n",
       "137       KNOWN           2     HIST1H1B      HIST1H1B  876888  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 2: Schizophrenia (SCZ) de novo mutation (DNM) data  \n",
    "Thoughts: Why did the authors remove \"start_lost\" information (one row of info)?  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "## get DNM data from one file\n",
    "dnm_path_s3 = Path.cwd().joinpath('supporting_files', 'SCZ_DNM', \\\n",
    "                                  '2014_Feb_De_novo_mutation_in_SCZ_Nature_623_trios_s3.txt')\n",
    "dnm = pd.read_table(dnm_path_s3)\n",
    "## note that the child phenoypes are ASD, control, ID, SZ, Unaffected sibling\n",
    "dnm = dnm[dnm['Child phenotype'] == 'SZ']  ## get DNM in individuals with schizophrenia  \n",
    "\n",
    "## get DNM from the other file\n",
    "dnm_path_s2 = Path.cwd().joinpath('supporting_files', 'SCZ_DNM', \\\n",
    "                                  '2014_Feb_De_novo_mutation_in_SCZ_Nature_623_trios_s2.txt')\n",
    "fromer = pd.read_table(dnm_path_s2)\n",
    "fromer['Child phenotype'] = np.repeat('SZ', len(fromer))\n",
    "fromer['Study'] = np.repeat('Fromer', len(fromer))\n",
    "## all of these are DNM in individuals with schizophrenia   \n",
    "\n",
    "dnm = dnm.merge(fromer, how = 'outer', on = ['Child phenotype', 'Study', 'Locus', 'Reference allele', \\\n",
    "                 'Alternate allele', 'Gene annotations', 'Genes', 'Transcripts', 'Transcript annotations', \\\n",
    "                 'Codon changes', 'Protein changes', 'Tri-nucleotide base context (forward strand) for SNV'])\n",
    "\n",
    "## There are 10 types of gene annotations. Authors kept 6: \"esplice\", \"frameshift\", \"nonsense\", \"missense\", \"codon-deletion\", \"code-insertion\". More efficient to take out the others\n",
    "removed_annots = ['silent', 'silent,missense', 'silent,silent', 'start-lost']\n",
    "dnm = dnm[~ dnm['Gene annotations'].isin(removed_annots)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Child phenotype</th>\n",
       "      <th>Study</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Reference allele</th>\n",
       "      <th>Alternate allele</th>\n",
       "      <th>Gene annotations</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Transcripts</th>\n",
       "      <th>Transcript annotations</th>\n",
       "      <th>Codon changes</th>\n",
       "      <th>...</th>\n",
       "      <th>Mother genotype meta-information</th>\n",
       "      <th>Sequencing center</th>\n",
       "      <th>Inferred parental origin of mutation</th>\n",
       "      <th>Proband ID</th>\n",
       "      <th>Proband gender</th>\n",
       "      <th>Family history</th>\n",
       "      <th>Paternal age at proband birth</th>\n",
       "      <th>Maternal age at proband birth</th>\n",
       "      <th>Proband age at onset of disease</th>\n",
       "      <th>Does the proband have a de novo CNV (Kirov, et al., 2012)?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SZ</td>\n",
       "      <td>Girard</td>\n",
       "      <td>chr19:36673714</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>ZNF565</td>\n",
       "      <td>NM_001042474,NM_152477</td>\n",
       "      <td>missense,missense</td>\n",
       "      <td>c.1154CAT&gt;CGT,c.1154CAT&gt;CGT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SZ</td>\n",
       "      <td>Girard</td>\n",
       "      <td>chr21:16338349</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>NRIP1</td>\n",
       "      <td>NM_003489</td>\n",
       "      <td>missense</td>\n",
       "      <td>c.2165AAA&gt;ACA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZ</td>\n",
       "      <td>Girard</td>\n",
       "      <td>chr12:57579450</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>NM_002332</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>c.6600TAC&gt;TAA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SZ</td>\n",
       "      <td>Girard</td>\n",
       "      <td>chr17:79637360</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>CCDC137</td>\n",
       "      <td>NM_199287</td>\n",
       "      <td>missense</td>\n",
       "      <td>c.374TAT&gt;TGT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SZ</td>\n",
       "      <td>Girard</td>\n",
       "      <td>chr3:122146472</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>KPNA1</td>\n",
       "      <td>NM_002264,NR_026698</td>\n",
       "      <td>nonsense,npcRNA</td>\n",
       "      <td>c.1342GAA&gt;TAA,.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Child phenotype   Study           Locus Reference allele Alternate allele  \\\n",
       "0              SZ  Girard  chr19:36673714                T                C   \n",
       "1              SZ  Girard  chr21:16338349                T                G   \n",
       "2              SZ  Girard  chr12:57579450                C                A   \n",
       "3              SZ  Girard  chr17:79637360                A                G   \n",
       "4              SZ  Girard  chr3:122146472                C                A   \n",
       "\n",
       "  Gene annotations    Genes             Transcripts Transcript annotations  \\\n",
       "0         missense   ZNF565  NM_001042474,NM_152477      missense,missense   \n",
       "1         missense    NRIP1               NM_003489               missense   \n",
       "2         nonsense     LRP1               NM_002332               nonsense   \n",
       "3         missense  CCDC137               NM_199287               missense   \n",
       "4         nonsense    KPNA1     NM_002264,NR_026698        nonsense,npcRNA   \n",
       "\n",
       "                 Codon changes  ... Mother genotype meta-information  \\\n",
       "0  c.1154CAT>CGT,c.1154CAT>CGT  ...                              NaN   \n",
       "1                c.2165AAA>ACA  ...                              NaN   \n",
       "2                c.6600TAC>TAA  ...                              NaN   \n",
       "3                 c.374TAT>TGT  ...                              NaN   \n",
       "4              c.1342GAA>TAA,.  ...                              NaN   \n",
       "\n",
       "  Sequencing center Inferred parental origin of mutation Proband ID  \\\n",
       "0               NaN                                  NaN        NaN   \n",
       "1               NaN                                  NaN        NaN   \n",
       "2               NaN                                  NaN        NaN   \n",
       "3               NaN                                  NaN        NaN   \n",
       "4               NaN                                  NaN        NaN   \n",
       "\n",
       "  Proband gender Family history Paternal age at proband birth  \\\n",
       "0            NaN            NaN                           NaN   \n",
       "1            NaN            NaN                           NaN   \n",
       "2            NaN            NaN                           NaN   \n",
       "3            NaN            NaN                           NaN   \n",
       "4            NaN            NaN                           NaN   \n",
       "\n",
       "  Maternal age at proband birth Proband age at onset of disease  \\\n",
       "0                           NaN                             NaN   \n",
       "1                           NaN                             NaN   \n",
       "2                           NaN                             NaN   \n",
       "3                           NaN                             NaN   \n",
       "4                           NaN                             NaN   \n",
       "\n",
       "  Does the proband have a de novo CNV (Kirov, et al., 2012)?  \n",
       "0                                                NaN          \n",
       "1                                                NaN          \n",
       "2                                                NaN          \n",
       "3                                                NaN          \n",
       "4                                                NaN          \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dnm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up \n",
    "del fromer, dnm_path_s2, dnm_path_s3, removed_annots\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 3: SCZ-specific differential expression (DE) data  \n",
    "This data is already in p-value form  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>MAPPED_genes</th>\n",
       "      <th>logFC</th>\n",
       "      <th>CI.L</th>\n",
       "      <th>CI.R</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000141433</td>\n",
       "      <td>ADCYAP1</td>\n",
       "      <td>0.35520</td>\n",
       "      <td>0.27200</td>\n",
       "      <td>0.43840</td>\n",
       "      <td>4.472</td>\n",
       "      <td>8.386</td>\n",
       "      <td>4.159000e-16</td>\n",
       "      <td>6.830000e-12</td>\n",
       "      <td>25.90230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000138640</td>\n",
       "      <td>FAM13A</td>\n",
       "      <td>-0.09541</td>\n",
       "      <td>-0.12470</td>\n",
       "      <td>-0.06606</td>\n",
       "      <td>6.365</td>\n",
       "      <td>-6.387</td>\n",
       "      <td>3.589000e-10</td>\n",
       "      <td>2.908000e-06</td>\n",
       "      <td>12.69002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000135299</td>\n",
       "      <td>ANKRD6</td>\n",
       "      <td>0.13390</td>\n",
       "      <td>0.09229</td>\n",
       "      <td>0.17550</td>\n",
       "      <td>5.291</td>\n",
       "      <td>6.322</td>\n",
       "      <td>5.312000e-10</td>\n",
       "      <td>2.908000e-06</td>\n",
       "      <td>12.35552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000104327</td>\n",
       "      <td>CALB1</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.13140</td>\n",
       "      <td>0.25290</td>\n",
       "      <td>5.241</td>\n",
       "      <td>6.217</td>\n",
       "      <td>9.951000e-10</td>\n",
       "      <td>4.085000e-06</td>\n",
       "      <td>11.75607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000125170</td>\n",
       "      <td>DOK4</td>\n",
       "      <td>0.10560</td>\n",
       "      <td>0.07147</td>\n",
       "      <td>0.13970</td>\n",
       "      <td>3.795</td>\n",
       "      <td>6.079</td>\n",
       "      <td>2.254000e-09</td>\n",
       "      <td>7.405000e-06</td>\n",
       "      <td>11.01800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             genes MAPPED_genes    logFC     CI.L     CI.R  AveExpr      t  \\\n",
       "0  ENSG00000141433      ADCYAP1  0.35520  0.27200  0.43840    4.472  8.386   \n",
       "1  ENSG00000138640       FAM13A -0.09541 -0.12470 -0.06606    6.365 -6.387   \n",
       "2  ENSG00000135299       ANKRD6  0.13390  0.09229  0.17550    5.291  6.322   \n",
       "3  ENSG00000104327        CALB1  0.19220  0.13140  0.25290    5.241  6.217   \n",
       "4  ENSG00000125170         DOK4  0.10560  0.07147  0.13970    3.795  6.079   \n",
       "\n",
       "        P.Value     adj.P.Val         B  \n",
       "0  4.159000e-16  6.830000e-12  25.90230  \n",
       "1  3.589000e-10  2.908000e-06  12.69002  \n",
       "2  5.312000e-10  2.908000e-06  12.35552  \n",
       "3  9.951000e-10  4.085000e-06  11.75607  \n",
       "4  2.254000e-09  7.405000e-06  11.01800  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_path = Path.cwd().joinpath('supporting_files', 'SCZ_DE', \\\n",
    "          'CMC_MSSM-Penn-Pitt_DLPFC_mRNA_IlluminaHiSeq2500_gene-adjustedSVA-differentialExpression-includeAncestry-DxSCZ-DE.tsv')\n",
    "dexpr = pd.read_table(de_path)\n",
    "\n",
    "dexpr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up \n",
    "del de_path\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 4-5: brain Hi-C data (CP + GZ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>ts_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "      <th>DTS</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185130.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>936538</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182611.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>929640</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196374.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>929425</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000233822.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>905924</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184357.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>876888</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP SNP_chr  SNP_pos_hg19             Name Description chrom  type  \\\n",
       "0  rs115329265    chr6      28712247  ENSG00000185130   HIST1H2BL  chr6  gene   \n",
       "1  rs115329265    chr6      28712247  ENSG00000182611   HIST1H2AJ  chr6  gene   \n",
       "2  rs115329265    chr6      28712247  ENSG00000196374   HIST1H2BM  chr6  gene   \n",
       "3  rs115329265    chr6      28712247  ENSG00000233822   HIST1H2BN  chr6  gene   \n",
       "4  rs115329265    chr6      28712247  ENSG00000184357    HIST1H1B  chr6  gene   \n",
       "\n",
       "   start_hg19  end_hg19 strand              ts_id       gene_type gene_status  \\\n",
       "0    27775257  27775709      -  ENSG00000185130.4  protein_coding       KNOWN   \n",
       "1    27782112  27782607      -  ENSG00000182611.3  protein_coding       KNOWN   \n",
       "2    27782822  27783267      +  ENSG00000196374.5  protein_coding       KNOWN   \n",
       "3    27806323  27819974      +  ENSG00000233822.2  protein_coding       KNOWN   \n",
       "4    27834570  27835359      -  ENSG00000184357.3  protein_coding       KNOWN   \n",
       "\n",
       "   loci_level alias_symbol official_name     DTS  cp_enhancer_no  \\\n",
       "0           2    HIST1H2BL     HIST1H2BL  936538              10   \n",
       "1           2    HIST1H2AJ     HIST1H2AJ  929640               8   \n",
       "2           2    HIST1H2BM     HIST1H2BM  929425               8   \n",
       "3           2    HIST1H2BN     HIST1H2BN  905924               6   \n",
       "4           2     HIST1H1B      HIST1H1B  876888               4   \n",
       "\n",
       "   gz_enhancer_no  \n",
       "0              10  \n",
       "1               9  \n",
       "2               9  \n",
       "3               7  \n",
       "4               5  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cp_path = Path.cwd().joinpath('supporting_files', 'BrainHiC', 'S22_TSS_CP.txt')\n",
    "gz_path = Path.cwd().joinpath('supporting_files', 'BrainHiC', 'S23_TSS_GZ.txt')\n",
    "\n",
    "cp = pd.read_table(cp_path)  ## cp has 221069 rows\n",
    "cp.drop(labels = 'Unnamed: 7', axis = 'columns', inplace=True)\n",
    "gz = pd.read_table(gz_path)  ## gz has 228323 rows\n",
    "\n",
    "## make a column of enhancer coordinate info that makes it easy to compare positions  \n",
    "cp['cp_enhancer_no'] = cp['chr'].str.cat(cp[['interacting_bin_start', 'interacting_bin_end']].astype(str), \\\n",
    "                  sep = ':')\n",
    "gz['gz_enhancer_no'] = gz['chr'].str.cat(gz[['interacting_bin_start', 'interacting_bin_end']].astype(str), \\\n",
    "                  sep = ':')\n",
    "\n",
    "## count how many unique enhancers (by chr, coordinates) per gene in dataset\n",
    "cp_counts = cp[['ENSGID_for_TSS', 'cp_enhancer_no']].groupby('ENSGID_for_TSS')['cp_enhancer_no'].nunique().to_frame()\n",
    "cp_counts.reset_index(inplace = True)  ## 34255 ENSG IDs/rows\n",
    "\n",
    "gz_counts = gz[['ENSGID_for_TSS', 'gz_enhancer_no']].groupby('ENSGID_for_TSS')['gz_enhancer_no'].nunique()\n",
    "  ## 34255 ENSG IDs/rows\n",
    "gz_counts = gz_counts.to_frame()\n",
    "gz_counts.reset_index(inplace = True)  ## 35019 ENSG IDs/rows\n",
    "\n",
    "## adds columns to candidate_gene df \n",
    "candidate_genes = candidate_genes.merge(cp_counts, how = 'left', left_on = 'Name', right_on = 'ENSGID_for_TSS')\n",
    "candidate_genes = candidate_genes.merge(gz_counts, how = 'left', left_on = 'Name', right_on = 'ENSGID_for_TSS')\n",
    "candidate_genes.drop(labels = ['ENSGID_for_TSS_x', 'ENSGID_for_TSS_y'], axis = 'columns', inplace=True)\n",
    "\n",
    "## replace NAs with 0 for cp_enhancer_no, gz_enhancer_no\n",
    "candidate_genes['cp_enhancer_no'].fillna(0, downcast='infer', inplace=True)\n",
    "candidate_genes['gz_enhancer_no'].fillna(0, downcast='infer', inplace=True)\n",
    "\n",
    "candidate_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up \n",
    "del cp_counts, gz_counts\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 6: capture Hi-C data (cell line, not specific to brain)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>ts_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "      <th>DTS</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185130.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>936538</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000182611.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>929640</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000196374.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>929425</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000233822.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>905924</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184357.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>876888</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>130.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP SNP_chr  SNP_pos_hg19             Name Description chrom  type  \\\n",
       "0  rs115329265    chr6      28712247  ENSG00000185130   HIST1H2BL  chr6  gene   \n",
       "1  rs115329265    chr6      28712247  ENSG00000182611   HIST1H2AJ  chr6  gene   \n",
       "2  rs115329265    chr6      28712247  ENSG00000196374   HIST1H2BM  chr6  gene   \n",
       "3  rs115329265    chr6      28712247  ENSG00000233822   HIST1H2BN  chr6  gene   \n",
       "4  rs115329265    chr6      28712247  ENSG00000184357    HIST1H1B  chr6  gene   \n",
       "\n",
       "   start_hg19  end_hg19 strand              ts_id       gene_type gene_status  \\\n",
       "0    27775257  27775709      -  ENSG00000185130.4  protein_coding       KNOWN   \n",
       "1    27782112  27782607      -  ENSG00000182611.3  protein_coding       KNOWN   \n",
       "2    27782822  27783267      +  ENSG00000196374.5  protein_coding       KNOWN   \n",
       "3    27806323  27819974      +  ENSG00000233822.2  protein_coding       KNOWN   \n",
       "4    27834570  27835359      -  ENSG00000184357.3  protein_coding       KNOWN   \n",
       "\n",
       "   loci_level alias_symbol official_name     DTS  cp_enhancer_no  \\\n",
       "0           2    HIST1H2BL     HIST1H2BL  936538              10   \n",
       "1           2    HIST1H2AJ     HIST1H2AJ  929640               8   \n",
       "2           2    HIST1H2BM     HIST1H2BM  929425               8   \n",
       "3           2    HIST1H2BN     HIST1H2BN  905924               6   \n",
       "4           2     HIST1H1B      HIST1H1B  876888               4   \n",
       "\n",
       "   gz_enhancer_no  cap4_enhancer_no  \n",
       "0              10             223.0  \n",
       "1               9               0.0  \n",
       "2               9             211.0  \n",
       "3               7              26.0  \n",
       "4               5             130.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "capHiC_path = Path.cwd().joinpath('supporting_files', 'capHiC', 'GM12878_DRE_number')\n",
    "capHiC = pd.read_table(capHiC_path)\n",
    "\n",
    "candidate_genes = candidate_genes.merge(capHiC[['Name', 'cap4_enhancer_no']], how = 'left', \\\n",
    "                      on = 'Name')\n",
    "\n",
    "candidate_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supporting evidence 7: FANTOM5 data \n",
    "I do get 66899 enhancer-promoter links at end of filtering FDR < 1 (mentioned in original paper)!    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "      <th>DTS</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>936538</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>223.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>929640</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>929425</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>211.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>905924</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>26.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>876888</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>130.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP SNP_chr  SNP_pos_hg19             Name Description chrom  type  \\\n",
       "0  rs115329265    chr6      28712247  ENSG00000185130   HIST1H2BL  chr6  gene   \n",
       "1  rs115329265    chr6      28712247  ENSG00000182611   HIST1H2AJ  chr6  gene   \n",
       "2  rs115329265    chr6      28712247  ENSG00000196374   HIST1H2BM  chr6  gene   \n",
       "3  rs115329265    chr6      28712247  ENSG00000233822   HIST1H2BN  chr6  gene   \n",
       "4  rs115329265    chr6      28712247  ENSG00000184357    HIST1H1B  chr6  gene   \n",
       "\n",
       "   start_hg19  end_hg19 strand  ...       gene_type gene_status loci_level  \\\n",
       "0    27775257  27775709      -  ...  protein_coding       KNOWN          2   \n",
       "1    27782112  27782607      -  ...  protein_coding       KNOWN          2   \n",
       "2    27782822  27783267      +  ...  protein_coding       KNOWN          2   \n",
       "3    27806323  27819974      +  ...  protein_coding       KNOWN          2   \n",
       "4    27834570  27835359      -  ...  protein_coding       KNOWN          2   \n",
       "\n",
       "   alias_symbol official_name     DTS  cp_enhancer_no  gz_enhancer_no  \\\n",
       "0     HIST1H2BL     HIST1H2BL  936538              10              10   \n",
       "1     HIST1H2AJ     HIST1H2AJ  929640               8               9   \n",
       "2     HIST1H2BM     HIST1H2BM  929425               8               9   \n",
       "3     HIST1H2BN     HIST1H2BN  905924               6               7   \n",
       "4      HIST1H1B      HIST1H1B  876888               4               5   \n",
       "\n",
       "   cap4_enhancer_no  fantom5_enhancer_no  \n",
       "0             223.0                    4  \n",
       "1               0.0                    4  \n",
       "2             211.0                    4  \n",
       "3              26.0                    6  \n",
       "4             130.0                    4  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FANTOM_path = Path.cwd().joinpath('supporting_files', 'Fantom5', 'enhancer_tss_associations.bed.txt')\n",
    "FANTOM5_data = pd.read_table(FANTOM_path)  ## 66942 rows\n",
    " \n",
    "## parsing name column: adds NA to 35 rows\n",
    "tempFANTOM = FANTOM5_data['name'].str.split(pat = ';|:', expand = True)  ## splits name column into many columns\n",
    "tempFANTOM.drop(labels = [4, 6], axis = 'columns', inplace=True)  ## drop Rstat/FDR strings\n",
    "tempFANTOM.columns = ['chr', 'pos_range', 'NCBI_ID', 'gene', 'Rstat', 'FDR']  ## assign columns\n",
    "FANTOM5_data = tempFANTOM.join(FANTOM5_data)\n",
    "\n",
    "## removes those 35 rows. have 66907 left\n",
    "FANTOM5_data.dropna(axis = \"index\", inplace = True)\n",
    "\n",
    "## select all genes with FDR < 1 \n",
    "FANTOM5_data['FDR'] = FANTOM5_data['FDR'].astype(np.float64)\n",
    "FANTOM5_data.query('FDR < 1', inplace = True)\n",
    "## removes 8 rows: left with 66899 rows\n",
    "\n",
    "## count how many unique enhancers (identified by position) per gene in dataset\n",
    "FANTOM5_enhancer_count = FANTOM5_data[['gene', 'pos_range']].rename(columns = {'pos_range':'fantom5_enhancer_no'})\n",
    "FANTOM5_enhancer_count = FANTOM5_enhancer_count.groupby('gene')['fantom5_enhancer_no'].nunique().to_frame()\n",
    "FANTOM5_enhancer_count.reset_index(inplace = True) \n",
    "## 11602 rows (unique gene symbols)\n",
    "\n",
    "## adds fantom5_enhancer_no column to table\n",
    "candidate_genes = candidate_genes.merge(FANTOM5_enhancer_count, how = 'left', left_on = 'alias_symbol', \\\n",
    "                                        right_on = 'gene')\n",
    "candidate_genes.drop(labels = 'gene', axis = 'columns', inplace=True)\n",
    "# candidate_genes[candidate_genes['fantom5_enhancer_no'].isna()].shape\n",
    "## 597 NAs were added so need to replace those with 0. \n",
    "candidate_genes['fantom5_enhancer_no'].fillna(0, downcast='infer', inplace=True)\n",
    "\n",
    "candidate_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## clean up \n",
    "del FANTOM_path, FANTOM5_enhancer_count, capHiC_path\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating and combining p-values\n",
    "\n",
    "**CHANGES from original code:**\n",
    "* replaced missing values in capHiC column with 0, to match how missing values were handled in other enhancer columns \n",
    "* kept surrogate Bayes factor for each gene-SNP pair (use SNP ID and alias_symbol to retrieve during Gibbs sampling)  \n",
    "\n",
    "Questions (also mentioned in Intro):  \n",
    "{? why did original code appear to keep only one \"extra_weight\" (surrogate Bayes factor, the first match) per unique gene/\"official_name\" value?}  \n",
    "{? The Mahalanobis transformation is supposed to create variables that follow the univariable Gaussian (normal) distribution but I don't think they do based on the EDA...}  \n",
    "{? Why is dnm probability calculated using a hard-coded number: 20,000?}  \n",
    "{? why get negative log of the final \"extra_weights\"? Maybe to get a continuous measure (https://lesslikely.com/statistics/s-values/), or to get bigger numbers when meta p-value is small?}  \n",
    "\n",
    "### Mahalanobis transformation (ZCA whitening)   \n",
    "\n",
    "* Done on DTS, enhancer columns of supporting data (capHiC, FANTOM5, brain-specific HiC)  \n",
    "* This creates uncorrelated variables, which can then be used to find p-values "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DTS</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alias_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C10orf95</th>\n",
       "      <td>396281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIPPLY2</th>\n",
       "      <td>282711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf233</th>\n",
       "      <td>851625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 DTS  cap4_enhancer_no  fantom5_enhancer_no  cp_enhancer_no  \\\n",
       "alias_symbol                                                                  \n",
       "C10orf95      396281               NaN                    0               2   \n",
       "RIPPLY2       282711               NaN                    1               0   \n",
       "C1orf233      851625               NaN                    0               5   \n",
       "\n",
       "              gz_enhancer_no  \n",
       "alias_symbol                  \n",
       "C10orf95                   2  \n",
       "RIPPLY2                    1  \n",
       "C1orf233                   6  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## process data to get mean-centered values  \n",
    "zca_cols = ['DTS', 'cap4_enhancer_no', 'fantom5_enhancer_no', 'cp_enhancer_no', 'gz_enhancer_no']\n",
    "\n",
    "extra_evidence = candidate_genes[zca_cols].copy()\n",
    "extra_evidence.index = candidate_genes['alias_symbol']  ## trying this to keep track of gene names\n",
    "\n",
    "## show rows with missing values\n",
    "extra_evidence[extra_evidence.isnull().any(axis=1)]\n",
    "\n",
    "extra_evidence['cap4_enhancer_no'].fillna(0, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that only one column has NA values: cap4_enhancer_no. This is because DTS is calculated from genomic coordinates and the other enhancer columns replaced their missing values with 0 (in the original R code, the default value for the columns is 0: see extra_evi.R).  \n",
    "\n",
    "The original R code replaces the missing values here, aka cap4_enhancer_no's missing values, with the median for the column. I want to treat this column the same as the other enhancer columns, so I replace the 3 missing values with 0. \n",
    "\n",
    "Next is calculating the Mahalanobis transformation.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "## first calculate covariance matrix, and SVD on the covariance matrix\n",
    "cov_extra = extra_evidence.cov().values  ## gives 5 x 5 (pairwise covariances) numpy ndarray\n",
    "\n",
    "svd_u, svd_diag, svd_vh = np.linalg.svd(cov_extra) \n",
    "## u and v are the same 5 x 5 matrix, diag = singular values\n",
    "\n",
    "## find 1 / sqrt () for each value in d (same as () ^ (-0.5)). Then make these values \n",
    "##      into a diagonal matrix = new_s_diag (values only on diagonal, 0 otherwise). \n",
    "new_s_diag = np.diag(v = svd_diag**(-0.5))\n",
    "## make new set of transformations: U * new_s_diag * V(transpose). \n",
    "##      Looks like \"making a new covariance matrix for a new set of variables\". \n",
    "svd_newtrans = svd_u @ new_s_diag @ svd_vh\n",
    "## @ is matrix multiplication, on 5 x 5 matrices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Mahalanobis transformation is done on mean-centered values.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## get mean-centered values \n",
    "extra_evidence = extra_evidence.apply(lambda x: x - x.mean(), axis = 0)  ## applies to each column\n",
    "\n",
    "## transforming mean-centered values  \n",
    "extra_evidence = (svd_newtrans @ extra_evidence.T).T\n",
    "## 5 x 5 matrix * 5 x 1732 matrix = 5 x 1732 matrix. \n",
    "## Then transpose to get it back in the same form as original df: 1732 genes x 5 variables. \n",
    "\n",
    "## the columns are \"decorrelated versions\" of the original variables\n",
    "##  give them the same column names\n",
    "extra_evidence.columns = zca_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DTS</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alias_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HIST1H2BL</th>\n",
       "      <td>1.423880</td>\n",
       "      <td>2.489901</td>\n",
       "      <td>-0.069589</td>\n",
       "      <td>0.579968</td>\n",
       "      <td>0.453113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H2AJ</th>\n",
       "      <td>1.401154</td>\n",
       "      <td>-1.219034</td>\n",
       "      <td>0.184902</td>\n",
       "      <td>0.183018</td>\n",
       "      <td>0.585704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H2BM</th>\n",
       "      <td>1.400452</td>\n",
       "      <td>2.290412</td>\n",
       "      <td>-0.048384</td>\n",
       "      <td>0.165730</td>\n",
       "      <td>0.551136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H2BN</th>\n",
       "      <td>1.323046</td>\n",
       "      <td>-0.789068</td>\n",
       "      <td>0.497234</td>\n",
       "      <td>-0.039871</td>\n",
       "      <td>0.365755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H1B</th>\n",
       "      <td>1.227420</td>\n",
       "      <td>0.942483</td>\n",
       "      <td>0.063548</td>\n",
       "      <td>-0.258647</td>\n",
       "      <td>0.140395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   DTS  cap4_enhancer_no  fantom5_enhancer_no  cp_enhancer_no  \\\n",
       "alias_symbol                                                                    \n",
       "HIST1H2BL     1.423880          2.489901            -0.069589        0.579968   \n",
       "HIST1H2AJ     1.401154         -1.219034             0.184902        0.183018   \n",
       "HIST1H2BM     1.400452          2.290412            -0.048384        0.165730   \n",
       "HIST1H2BN     1.323046         -0.789068             0.497234       -0.039871   \n",
       "HIST1H1B      1.227420          0.942483             0.063548       -0.258647   \n",
       "\n",
       "              gz_enhancer_no  \n",
       "alias_symbol                  \n",
       "HIST1H2BL           0.453113  \n",
       "HIST1H2AJ           0.585704  \n",
       "HIST1H2BM           0.551136  \n",
       "HIST1H2BN           0.365755  \n",
       "HIST1H1B            0.140395  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extra_evidence.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the current setup, we would need to calculate the p-value for the lower tail for the DTS variable and the p-value for the upper tail for the other variables. This is because...  \n",
    "* DTS: the smaller the number (distance to SNP < expected), the more likely it is to be a disease gene (higher weight)  \n",
    "* for the others: the larger the number (number of DRE-promoter links > expected), the more likely it is to be a disease gene (higher weight). Recall the assumption made (Description of iRIGS above).    \n",
    "\n",
    "For the other columns, we flip their signs to flip the tails. Now for all variables, the smaller the number, the more likely it is to be a disease gene.  \n",
    "\n",
    "Note: the Mahalanobis transformation is supposed to create variables that follow the univariable Gaussian (normal) distribution but I feel like it doesn't...see the last code chunk for this section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "larger_likely_disease = ['cap4_enhancer_no', 'fantom5_enhancer_no', 'cp_enhancer_no', 'gz_enhancer_no']\n",
    "\n",
    "extra_evidence = extra_evidence.apply(lambda x: -x if x.name in larger_likely_disease else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NormaltestResult(statistic=array([ 399.3259364 ,  128.09458355, 1887.39056917,  818.13328282,\n",
       "        650.04895805]), pvalue=array([1.93853957e-087, 1.52972936e-028, 0.00000000e+000, 2.21113319e-178,\n",
       "       6.97689643e-142]))"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## notice how small the p-values for the normality tests are \n",
    "## ref: https://docs.scipy.org/doc/scipy/reference/tutorial/stats.html#special-tests-for-normal-distributions\n",
    "stats.normaltest(extra_evidence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make p-values from transformed variables  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ref for stats.norm.cdf being the same as R's pnorm : https://stackoverflow.com/a/42065440\n",
    "extra_evidence = extra_evidence.apply(lambda x: stats.norm.cdf(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del larger_likely_disease, zca_cols, cov_extra, svd_u, svd_diag, svd_vh, new_s_diag, svd_newtrans\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add p-values for DNM and DE data \n",
    "\n",
    "Add and process p-values from DNM (de-novo mutation) and DE (differential expression) data. \n",
    "\n",
    "Notice:\n",
    "* This data is schizophrenia specific.  \n",
    "* The hard-coding denominator of 20,000 for the DNM probability calculation. Perhaps can be replaced with a different number?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DTS</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "      <th>dnm</th>\n",
       "      <th>DE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.922759</td>\n",
       "      <td>0.006389</td>\n",
       "      <td>0.527739</td>\n",
       "      <td>0.280968</td>\n",
       "      <td>0.325234</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.086510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.919416</td>\n",
       "      <td>0.888584</td>\n",
       "      <td>0.426653</td>\n",
       "      <td>0.427392</td>\n",
       "      <td>0.279037</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.919311</td>\n",
       "      <td>0.010999</td>\n",
       "      <td>0.519295</td>\n",
       "      <td>0.434185</td>\n",
       "      <td>0.290770</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.907090</td>\n",
       "      <td>0.784964</td>\n",
       "      <td>0.309512</td>\n",
       "      <td>0.515902</td>\n",
       "      <td>0.357274</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.006472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.890168</td>\n",
       "      <td>0.172973</td>\n",
       "      <td>0.474665</td>\n",
       "      <td>0.602046</td>\n",
       "      <td>0.444174</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        DTS  cap4_enhancer_no  fantom5_enhancer_no  cp_enhancer_no  \\\n",
       "0  0.922759          0.006389             0.527739        0.280968   \n",
       "1  0.919416          0.888584             0.426653        0.427392   \n",
       "2  0.919311          0.010999             0.519295        0.434185   \n",
       "3  0.907090          0.784964             0.309512        0.515902   \n",
       "4  0.890168          0.172973             0.474665        0.602046   \n",
       "\n",
       "   gz_enhancer_no     dnm        DE  \n",
       "0        0.325234  0.9666  0.086510  \n",
       "1        0.279037  0.9666       NaN  \n",
       "2        0.290770  0.9666       NaN  \n",
       "3        0.357274  0.9666  0.006472  \n",
       "4        0.444174  0.9666       NaN  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## DNM data \n",
    "dnm_prob = dnm['Genes'].nunique() / 20000\n",
    "\n",
    "## only 60 candidate genes have schizophrenia-associated DNM: they are assigned the dnm_prob\n",
    "# len(extra_evidence[extra_evidence.index.isin(dnm['Genes'])])\n",
    "extra_evidence['dnm'] = [dnm_prob if x in dnm['Genes'].to_list() \\\n",
    "                                  else (1 - dnm_prob) \\\n",
    "                         for x in extra_evidence.index]\n",
    "\n",
    "## DE data \n",
    "## note that extra_evidence df has same row order as candidate_genes\n",
    "## add column of ENSG IDs to make merge possible with dexpr\n",
    "extra_evidence.reset_index(drop = True, inplace = True)  ## won't work unless gene symbol index is dropped\n",
    "extra_evidence['Name'] = candidate_genes['Name']\n",
    "\n",
    "## merge with dexpr\n",
    "extra_evidence = extra_evidence.merge(dexpr[['genes','P.Value']], \\\n",
    "                                      how = 'left', left_on = 'Name', right_on = 'genes')\n",
    "## clean-up dataframe\n",
    "extra_evidence.drop(labels = ['Name', 'genes'], axis = 'columns', inplace = True)\n",
    "extra_evidence.rename(columns={'P.Value': 'DE'}, inplace = True)\n",
    "extra_evidence.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Missing DE p-values should be replaced with the median p-value, rather than giving it a p-value of 0 (aka significant). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1261"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The median DE p-value is 0.3433\n"
     ]
    }
   ],
   "source": [
    "## 1261 genes have DE info, 376 have missing values \n",
    "extra_evidence['DE'].count()  ## method to get number of non-missing values in the column\n",
    "print(\"The median DE p-value is\", extra_evidence['DE'].median(skipna = True))\n",
    "\n",
    "## replace missing values with median\n",
    "extra_evidence['DE'].fillna(extra_evidence['DE'].median(skipna = True), inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DTS</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "      <th>dnm</th>\n",
       "      <th>DE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.922759</td>\n",
       "      <td>0.006389</td>\n",
       "      <td>0.527739</td>\n",
       "      <td>0.280968</td>\n",
       "      <td>0.325234</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.086510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.919416</td>\n",
       "      <td>0.888584</td>\n",
       "      <td>0.426653</td>\n",
       "      <td>0.427392</td>\n",
       "      <td>0.279037</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.343300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.919311</td>\n",
       "      <td>0.010999</td>\n",
       "      <td>0.519295</td>\n",
       "      <td>0.434185</td>\n",
       "      <td>0.290770</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.343300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.907090</td>\n",
       "      <td>0.784964</td>\n",
       "      <td>0.309512</td>\n",
       "      <td>0.515902</td>\n",
       "      <td>0.357274</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.006472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.890168</td>\n",
       "      <td>0.172973</td>\n",
       "      <td>0.474665</td>\n",
       "      <td>0.602046</td>\n",
       "      <td>0.444174</td>\n",
       "      <td>0.9666</td>\n",
       "      <td>0.343300</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        DTS  cap4_enhancer_no  fantom5_enhancer_no  cp_enhancer_no  \\\n",
       "0  0.922759          0.006389             0.527739        0.280968   \n",
       "1  0.919416          0.888584             0.426653        0.427392   \n",
       "2  0.919311          0.010999             0.519295        0.434185   \n",
       "3  0.907090          0.784964             0.309512        0.515902   \n",
       "4  0.890168          0.172973             0.474665        0.602046   \n",
       "\n",
       "   gz_enhancer_no     dnm        DE  \n",
       "0        0.325234  0.9666  0.086510  \n",
       "1        0.279037  0.9666  0.343300  \n",
       "2        0.290770  0.9666  0.343300  \n",
       "3        0.357274  0.9666  0.006472  \n",
       "4        0.444174  0.9666  0.343300  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extra_evidence.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fisher's product method of combining data  \n",
    "See: https://en.wikipedia.org/wiki/Fisher%27s_method\n",
    "\n",
    "When individual p-values are small, the meta-test-statistic should be large. When all individual null hypotheses are true and their test stats/p-values are independent, this meta-test-statistic has a chi-squared dist with 2k degrees of freedom (k = number of tests that are combined).  \n",
    "\n",
    "The meta p-values are then calculated using what the paper's supplement (pgs. 2-3) describes as the \"inverse cumulative distribution function (CDF) of chi-squared (chi2) distribution.\" In practice, this is 1 - (chi2's CDF). In Python, this is found using scipy's survival function (sf). See methods in https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html\n",
    "\n",
    "**The final values here are used as surrogates for the Bayesian factor!** This is explained in the Supplementary Note for the iRIGS paper.     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "## From Fisher's method: chi-squared dist has degrees of freedom = 2 * number of tests  \n",
    "fishers_chi_df = 2 * len(extra_evidence.columns)\n",
    "\n",
    "## np.log default IS A NATURAL LOG \n",
    "## this is the meta-test-statistic. sum is across the row \n",
    "fishers_combo = extra_evidence.apply(lambda x: -2 * np.sum(np.log(x)), axis = 1)  \n",
    "## this is a pandas Series object\n",
    "\n",
    "## get p-values of the meta-test-statistic:  \n",
    "fishers_combo = stats.chi2.sf(fishers_combo, df = fishers_chi_df)\n",
    "\n",
    "## get negative log of all these weights. I think this is so larger weights -> \n",
    "##     the gene is more likely to be important for the SNP GWAS signal. \n",
    "fishers_combo = -np.log(fishers_combo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_chr</th>\n",
       "      <th>SNP_pos_hg19</th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start_hg19</th>\n",
       "      <th>end_hg19</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>loci_level</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>official_name</th>\n",
       "      <th>DTS</th>\n",
       "      <th>cp_enhancer_no</th>\n",
       "      <th>gz_enhancer_no</th>\n",
       "      <th>cap4_enhancer_no</th>\n",
       "      <th>fantom5_enhancer_no</th>\n",
       "      <th>transformed_Fisherprod</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000185130</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27775257</td>\n",
       "      <td>27775709</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>936538</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>223.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2.360891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000182611</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782112</td>\n",
       "      <td>27782607</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>929640</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.153445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000196374</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27782822</td>\n",
       "      <td>27783267</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>929425</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>211.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.329594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27806323</td>\n",
       "      <td>27819974</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>905924</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>26.0</td>\n",
       "      <td>6</td>\n",
       "      <td>1.270672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28712247</td>\n",
       "      <td>ENSG00000184357</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>gene</td>\n",
       "      <td>27834570</td>\n",
       "      <td>27835359</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>876888</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>130.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.278950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140143664</td>\n",
       "      <td>ENSG00000178913</td>\n",
       "      <td>TAF7</td>\n",
       "      <td>chr5</td>\n",
       "      <td>gene</td>\n",
       "      <td>140698057</td>\n",
       "      <td>140700330</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>TAF7</td>\n",
       "      <td>TAF7</td>\n",
       "      <td>556666</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>84.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.044525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1633</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140143664</td>\n",
       "      <td>ENSG00000131504</td>\n",
       "      <td>DIAPH1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>gene</td>\n",
       "      <td>140894583</td>\n",
       "      <td>140998622</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>DIAPH1</td>\n",
       "      <td>DIAPH1</td>\n",
       "      <td>854958</td>\n",
       "      <td>30</td>\n",
       "      <td>25</td>\n",
       "      <td>186.0</td>\n",
       "      <td>8</td>\n",
       "      <td>5.518480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140143664</td>\n",
       "      <td>ENSG00000171720</td>\n",
       "      <td>HDAC3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>gene</td>\n",
       "      <td>141000443</td>\n",
       "      <td>141016437</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>HDAC3</td>\n",
       "      <td>HDAC3</td>\n",
       "      <td>872773</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140143664</td>\n",
       "      <td>ENSG00000120318</td>\n",
       "      <td>ARAP3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>gene</td>\n",
       "      <td>141013251</td>\n",
       "      <td>141061788</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>1</td>\n",
       "      <td>ARAP3</td>\n",
       "      <td>ARAP3</td>\n",
       "      <td>918124</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>69.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.824276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140143664</td>\n",
       "      <td>ENSG00000164620</td>\n",
       "      <td>RELL2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>gene</td>\n",
       "      <td>141016517</td>\n",
       "      <td>141020644</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>2</td>\n",
       "      <td>RELL2</td>\n",
       "      <td>RELL2</td>\n",
       "      <td>872853</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.026325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1637 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNP SNP_chr  SNP_pos_hg19             Name Description  \\\n",
       "0          rs115329265    chr6      28712247  ENSG00000185130   HIST1H2BL   \n",
       "1          rs115329265    chr6      28712247  ENSG00000182611   HIST1H2AJ   \n",
       "2          rs115329265    chr6      28712247  ENSG00000196374   HIST1H2BM   \n",
       "3          rs115329265    chr6      28712247  ENSG00000233822   HIST1H2BN   \n",
       "4          rs115329265    chr6      28712247  ENSG00000184357    HIST1H1B   \n",
       "...                ...     ...           ...              ...         ...   \n",
       "1632  chr5_140143664_I    chr5     140143664  ENSG00000178913        TAF7   \n",
       "1633  chr5_140143664_I    chr5     140143664  ENSG00000131504      DIAPH1   \n",
       "1634  chr5_140143664_I    chr5     140143664  ENSG00000171720       HDAC3   \n",
       "1635  chr5_140143664_I    chr5     140143664  ENSG00000120318       ARAP3   \n",
       "1636  chr5_140143664_I    chr5     140143664  ENSG00000164620       RELL2   \n",
       "\n",
       "     chrom  type  start_hg19   end_hg19 strand  ... gene_status loci_level  \\\n",
       "0     chr6  gene    27775257   27775709      -  ...       KNOWN          2   \n",
       "1     chr6  gene    27782112   27782607      -  ...       KNOWN          2   \n",
       "2     chr6  gene    27782822   27783267      +  ...       KNOWN          2   \n",
       "3     chr6  gene    27806323   27819974      +  ...       KNOWN          2   \n",
       "4     chr6  gene    27834570   27835359      -  ...       KNOWN          2   \n",
       "...    ...   ...         ...        ...    ...  ...         ...        ...   \n",
       "1632  chr5  gene   140698057  140700330      -  ...       KNOWN          2   \n",
       "1633  chr5  gene   140894583  140998622      -  ...       KNOWN          2   \n",
       "1634  chr5  gene   141000443  141016437      -  ...       KNOWN          2   \n",
       "1635  chr5  gene   141013251  141061788      -  ...       KNOWN          1   \n",
       "1636  chr5  gene   141016517  141020644      +  ...       KNOWN          2   \n",
       "\n",
       "     alias_symbol  official_name     DTS cp_enhancer_no  gz_enhancer_no  \\\n",
       "0       HIST1H2BL      HIST1H2BL  936538             10              10   \n",
       "1       HIST1H2AJ      HIST1H2AJ  929640              8               9   \n",
       "2       HIST1H2BM      HIST1H2BM  929425              8               9   \n",
       "3       HIST1H2BN      HIST1H2BN  905924              6               7   \n",
       "4        HIST1H1B       HIST1H1B  876888              4               5   \n",
       "...           ...            ...     ...            ...             ...   \n",
       "1632         TAF7           TAF7  556666              3               4   \n",
       "1633       DIAPH1         DIAPH1  854958             30              25   \n",
       "1634        HDAC3          HDAC3  872773              1               0   \n",
       "1635        ARAP3          ARAP3  918124             16              20   \n",
       "1636        RELL2          RELL2  872853              0               0   \n",
       "\n",
       "      cap4_enhancer_no  fantom5_enhancer_no  transformed_Fisherprod  \n",
       "0                223.0                    4                2.360891  \n",
       "1                  0.0                    4                0.153445  \n",
       "2                211.0                    4                1.329594  \n",
       "3                 26.0                    6                1.270672  \n",
       "4                130.0                    4                0.278950  \n",
       "...                ...                  ...                     ...  \n",
       "1632              84.0                    2                0.044525  \n",
       "1633             186.0                    8                5.518480  \n",
       "1634              83.0                    0                0.008950  \n",
       "1635              69.0                    2                0.824276  \n",
       "1636              83.0                    5                0.026325  \n",
       "\n",
       "[1637 rows x 22 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_genes = candidate_genes.assign(transformed_Fisherprod = fishers_combo)\n",
    "candidate_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Optional code chunk: not required to run**\n",
    "\n",
    "I can explore what's going on with Fisher's method and check my manual calculations with the scipy stats function: https://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.stats.combine_pvalues.html "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 126, 1490, 1493]),)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "SNP                          chr7_2025096_I\n",
       "SNP_chr                                chr7\n",
       "SNP_pos_hg19                        2025096\n",
       "Name                        ENSG00000002822\n",
       "Description                          MAD1L1\n",
       "chrom                                  chr7\n",
       "type                                   gene\n",
       "start_hg19                          1855383\n",
       "end_hg19                            2272878\n",
       "strand                                    -\n",
       "ts_id                     ENSG00000002822.9\n",
       "gene_type                    protein_coding\n",
       "gene_status                           KNOWN\n",
       "loci_level                                1\n",
       "alias_symbol                         MAD1L1\n",
       "official_name                        MAD1L1\n",
       "DTS                                  247782\n",
       "cp_enhancer_no                           80\n",
       "gz_enhancer_no                           76\n",
       "cap4_enhancer_no                        112\n",
       "fantom5_enhancer_no                       1\n",
       "transformed_Fisherprod              52.6165\n",
       "Name: 126, dtype: object"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "DTS                    1.991888e-01\n",
       "cap4_enhancer_no       2.714492e-01\n",
       "fantom5_enhancer_no    7.579500e-01\n",
       "cp_enhancer_no         1.472854e-18\n",
       "gz_enhancer_no         5.766056e-12\n",
       "dnm                    9.666000e-01\n",
       "DE                     2.026000e-01\n",
       "Name: 126, dtype: float64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(143.52694878077799, 1.4091035662651953e-23)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## comparing the manual Fisher's method above with the scipy function for Fisher's method\n",
    "# fishers_combo[0:10] \n",
    "\n",
    "# list_fisherscombo = []\n",
    "# for ele in extra_evidence.values:\n",
    "#     list_fisherscombo.append(stats.combine_pvalues(pvalues = ele, method = 'fisher'))\n",
    "\n",
    "# for (test, value) in list_fisherscombo[0:10]:\n",
    "#     print(-np.log(value))\n",
    "\n",
    "## what had a very high surrogate Bayes factor? = a likely risk gene \n",
    "# np.argmax(fishers_combo)\n",
    "# candidate_genes.loc[1493,]\n",
    "# ## there were a lot of enhancers found and a smaller distance, surrogate Bayes factor is ~ 66.9\n",
    "# extra_evidence.loc[1493,]\n",
    "# ## one of the individual p-values is very low\n",
    "# stats.combine_pvalues(pvalues = extra_evidence.loc[1493,], method = 'fisher')\n",
    "# ## test stat is large\n",
    "\n",
    "# ## what had a very low surrogate Bayes factor? = not likely risk gene\n",
    "# np.argmin(fishers_combo)\n",
    "# candidate_genes.loc[1037,]\n",
    "# ## almost no enhancers found, distance from SNP is far, surrogate Bayes factor is ~ 0.0007 \n",
    "# extra_evidence.loc[1037,]\n",
    "# ## all individual p-values are high\n",
    "# stats.combine_pvalues(pvalues = extra_evidence.loc[1037,], method = 'fisher')\n",
    "# ## test stat is small\n",
    "\n",
    "# ## another with a very high surrogate Bayes factor\n",
    "np.where(fishers_combo >= 50)\n",
    "candidate_genes.loc[126,]\n",
    "## close distance, most enhancer numbers are high -> surrogate Bayes factor ~ 52.6\n",
    "extra_evidence.loc[126,]\n",
    "## two p-values are very small  \n",
    "stats.combine_pvalues(pvalues = extra_evidence.loc[126,], method = 'fisher')\n",
    "## test stat is high"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gibbs sampling: burn-in step\n",
    "**CHANGES from original code:**\n",
    "\n",
    "* Gibbs sampling is not done on SNP loci with only one candidate gene. This is a change from the original code, which would add a sample identical to the previous sample.  \n",
    "* Python version counts genes after every Gibbs sample in a table column. The R version originally stored rows of strings (genes) for every sample and counted genes only after a full round through SNP loci finished. \n",
    "\n",
    "**Thoughts:** \n",
    "* I think we work with square-roots of the threshold and sum of squared differences, so (1) sign/direction of difference won't matter with squaring and (2) we avoid dealing with very small numbers (loss of precision). This way of doing things was in the original R code.    \n",
    "* should SNPs with only one candidate gene be removed completely from the candidate gene space for Gibbs sampling? They will bias results because they are always used to calculate the network factors for other loci candidates - however, this may be what we want if we suspect that these candidate genes are highly likely to be disease risk genes.\n",
    "* what should happen in the unlikely scenario that all candidate genes for a SNP are also candidates for other loci (so there will be an empty table for calculating the network factor)? Perhaps we would need to go forward with a new random sample? Or would this \"never\" happen because users will not choose SNPs close together?\n",
    "* Would Python dictionaries (SNPs to genes) of dictionaries (each gene has a count) be a good data structure to use? \n",
    "\n",
    "[How to time chunks of code](https://stackoverflow.com/questions/2866380/how-can-i-time-a-code-segment-for-testing-performance-with-pythons-timeit#2866456)\n",
    "\n",
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>alias_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs115329265</th>\n",
       "      <th>20</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>OR2B3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1702294</th>\n",
       "      <th>37</th>\n",
       "      <td>rs1702294</td>\n",
       "      <td>DPYD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs11191419</th>\n",
       "      <th>59</th>\n",
       "      <td>rs11191419</td>\n",
       "      <td>CNNM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2007044</th>\n",
       "      <th>83</th>\n",
       "      <td>rs2007044</td>\n",
       "      <td>TULP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4129585</th>\n",
       "      <th>89</th>\n",
       "      <td>rs4129585</td>\n",
       "      <td>TSNARE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4240748</th>\n",
       "      <th>1500</th>\n",
       "      <td>rs4240748</td>\n",
       "      <td>CLLU1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2909457</th>\n",
       "      <th>1507</th>\n",
       "      <td>rs2909457</td>\n",
       "      <td>DPP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs56873913</th>\n",
       "      <th>1552</th>\n",
       "      <td>rs56873913</td>\n",
       "      <td>DKKL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10860964</th>\n",
       "      <th>1604</th>\n",
       "      <td>rs10860964</td>\n",
       "      <td>C12orf73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5_140143664_I</th>\n",
       "      <th>1619</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>SRA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    SNP alias_symbol\n",
       "SNP                                                 \n",
       "rs115329265      20         rs115329265        OR2B3\n",
       "rs1702294        37           rs1702294         DPYD\n",
       "rs11191419       59          rs11191419        CNNM2\n",
       "rs2007044        83           rs2007044        TULP3\n",
       "rs4129585        89           rs4129585      TSNARE1\n",
       "...                                 ...          ...\n",
       "rs4240748        1500         rs4240748        CLLU1\n",
       "rs2909457        1507         rs2909457         DPP4\n",
       "rs56873913       1552        rs56873913        DKKL1\n",
       "rs10860964       1604        rs10860964     C12orf73\n",
       "chr5_140143664_I 1619  chr5_140143664_I         SRA1\n",
       "\n",
       "[107 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>key</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>rs115329265,HIST1H2BL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>rs115329265,HIST1H2AJ</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>rs115329265,HIST1H2BM</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>rs115329265,HIST1H2BN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>rs115329265,HIST1H1B</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>TAF7</td>\n",
       "      <td>chr5_140143664_I,TAF7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1633</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>DIAPH1</td>\n",
       "      <td>chr5_140143664_I,DIAPH1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>HDAC3</td>\n",
       "      <td>chr5_140143664_I,HDAC3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>ARAP3</td>\n",
       "      <td>chr5_140143664_I,ARAP3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>RELL2</td>\n",
       "      <td>chr5_140143664_I,RELL2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1637 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNP alias_symbol                      key  counts\n",
       "0          rs115329265    HIST1H2BL    rs115329265,HIST1H2BL       0\n",
       "1          rs115329265    HIST1H2AJ    rs115329265,HIST1H2AJ       0\n",
       "2          rs115329265    HIST1H2BM    rs115329265,HIST1H2BM       0\n",
       "3          rs115329265    HIST1H2BN    rs115329265,HIST1H2BN       0\n",
       "4          rs115329265     HIST1H1B     rs115329265,HIST1H1B       0\n",
       "...                ...          ...                      ...     ...\n",
       "1632  chr5_140143664_I         TAF7    chr5_140143664_I,TAF7       0\n",
       "1633  chr5_140143664_I       DIAPH1  chr5_140143664_I,DIAPH1       0\n",
       "1634  chr5_140143664_I        HDAC3   chr5_140143664_I,HDAC3       0\n",
       "1635  chr5_140143664_I        ARAP3   chr5_140143664_I,ARAP3       0\n",
       "1636  chr5_140143664_I        RELL2   chr5_140143664_I,RELL2       0\n",
       "\n",
       "[1637 rows x 4 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "101"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## flag to remove candidate genes (rows) from network/columns calculations \n",
    "flag_exclude_sharedrowcols = True  \n",
    "\n",
    "## square-root of the threshold we want! stopping criterion\n",
    "stopping_threshold = math.sqrt(10**(-4))\n",
    "dif = stopping_threshold + 1  ## used for stopping criterion\n",
    "\n",
    "## round = a set of sampling thru all loci\n",
    "burnin_rounds = 0  ## counter for number of rounds \n",
    "max_burnin_rounds = 3000  ## maximum number of rounds allowed at burn-in step\n",
    "\n",
    "## used to record counts of SNP-gene pairs from all previous rounds\n",
    "burnin_gene_counts = candidate_genes[['SNP', 'alias_symbol']].copy()\n",
    "\n",
    "## get SNP loci with only one candidate gene: no need to use Gibbs sampling for these loci\n",
    "##     to find the most probable candidate\n",
    "## however, this locus is included in Gibbs sampling so it can be used to calculate the network factor\n",
    "SNPs_one_candidate = burnin_gene_counts.copy().groupby('SNP', sort = False).count()\n",
    "SNPs_one_candidate = set(SNPs_one_candidate[SNPs_one_candidate['alias_symbol'] == 1].index.to_list())\n",
    "\n",
    "## starting set: 1 per SNP locus \n",
    "## another way to do this: random sort and then grab first row for each SNP: \n",
    "##       https://stackoverflow.com/questions/36390406/sample-each-group-after-pandas-groupby\n",
    "current_geneset = burnin_gene_counts.copy().groupby('SNP', sort = False).apply(pd.DataFrame.sample, n=1)\n",
    "current_geneset\n",
    "\n",
    "## used to iterate through SNPs in the same order as current_gene_set's rows\n",
    "SNP_loci = burnin_gene_counts['SNP'].unique()  ## this is a numpy array\n",
    "\n",
    "## use key for fast updating of counts: \n",
    "##     using the dataframe index was messy because the current_geneset wouldn't have all the index values\n",
    "burnin_gene_counts['key'] = burnin_gene_counts['SNP'] + \",\" + burnin_gene_counts['alias_symbol']\n",
    "\n",
    "## set up counter for candidate genes\n",
    "burnin_gene_counts['counts'] = 0\n",
    "burnin_gene_counts\n",
    "\n",
    "## how many SNP loci are sampled per round\n",
    "SNPs_per_round = len(SNP_loci) - len(SNPs_one_candidate)\n",
    "SNPs_per_round"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Burn-in Loop\n",
    "\n",
    "Starting with the second round of Gibbs sampling, it prints difference from previous round. \n",
    "\n",
    "It took me 738 seconds (12.3 minutes) to do 619 rounds to reach a difference below threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 4.602453708534399 \n",
      "\n",
      "dif is 2.6366611143350505 \n",
      "\n",
      "dif is 1.7967273745811914 \n",
      "\n",
      "dif is 1.4965673234283758 \n",
      "\n",
      "dif is 1.1565741850795737 \n",
      "\n",
      "dif is 0.9401945199956643 \n",
      "\n",
      "dif is 0.8572687036211708 \n",
      "\n",
      "dif is 0.7288495283337117 \n",
      "\n",
      "dif is 0.6746137230919929 \n",
      "\n",
      "dif is 0.6213502768749306 \n",
      "\n",
      "dif is 0.5870879776592086 \n",
      "\n",
      "dif is 0.5553887625666095 \n",
      "\n",
      "dif is 0.4839013426032615 \n",
      "\n",
      "dif is 0.43745207841980227 \n",
      "\n",
      "dif is 0.40866478644892146 \n",
      "\n",
      "dif is 0.35881022128570367 \n",
      "\n",
      "dif is 0.3761855464144371 \n",
      "\n",
      "dif is 0.35153463660438916 \n",
      "\n",
      "dif is 0.33184940941941254 \n",
      "\n",
      "dif is 0.3125238106728691 \n",
      "\n",
      "dif is 0.29711965041905997 \n",
      "\n",
      "dif is 0.2905986410367301 \n",
      "\n",
      "dif is 0.2836970824377936 \n",
      "\n",
      "dif is 0.2547773889770668 \n",
      "\n",
      "dif is 0.2549454467288552 \n",
      "\n",
      "dif is 0.23911978693840166 \n",
      "\n",
      "dif is 0.2418581162166099 \n",
      "\n",
      "dif is 0.22712233106312585 \n",
      "\n",
      "dif is 0.21977091363272236 \n",
      "\n",
      "dif is 0.2209989456180215 \n",
      "\n",
      "dif is 0.21256829386221351 \n",
      "\n",
      "dif is 0.19099499314331592 \n",
      "\n",
      "dif is 0.19915729892938983 \n",
      "\n",
      "dif is 0.18179231203037274 \n",
      "\n",
      "dif is 0.17997660660069564 \n",
      "\n",
      "dif is 0.17141983247895531 \n",
      "\n",
      "dif is 0.1765821427140314 \n",
      "\n",
      "dif is 0.16724877180929093 \n",
      "\n",
      "dif is 0.16884326860141283 \n",
      "\n",
      "dif is 0.1642013488177643 \n",
      "\n",
      "dif is 0.16006042894579514 \n",
      "\n",
      "dif is 0.15860087662994743 \n",
      "\n",
      "dif is 0.15110695783174638 \n",
      "\n",
      "dif is 0.1427905691053636 \n",
      "\n",
      "dif is 0.13785894232429613 \n",
      "\n",
      "dif is 0.14131503611527424 \n",
      "\n",
      "dif is 0.13127412055091447 \n",
      "\n",
      "dif is 0.13660500713931403 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 50 \n",
      "\n",
      "dif is 0.1313661550293711 \n",
      "\n",
      "dif is 0.14000040235853517 \n",
      "\n",
      "dif is 0.12341093583293576 \n",
      "\n",
      "dif is 0.11909307149639416 \n",
      "\n",
      "dif is 0.11806430224093276 \n",
      "\n",
      "dif is 0.1122833864584971 \n",
      "\n",
      "dif is 0.11099061880102208 \n",
      "\n",
      "dif is 0.10991040664135743 \n",
      "\n",
      "dif is 0.11341228519930338 \n",
      "\n",
      "dif is 0.111153370065989 \n",
      "\n",
      "dif is 0.1058882838529758 \n",
      "\n",
      "dif is 0.10451034340546433 \n",
      "\n",
      "dif is 0.10288816168213069 \n",
      "\n",
      "dif is 0.09965426785024403 \n",
      "\n",
      "dif is 0.10118741522182359 \n",
      "\n",
      "dif is 0.10280374540572856 \n",
      "\n",
      "dif is 0.09951036217984838 \n",
      "\n",
      "dif is 0.09563856297988692 \n",
      "\n",
      "dif is 0.09686515664477727 \n",
      "\n",
      "dif is 0.09715429139862199 \n",
      "\n",
      "dif is 0.09148297540031583 \n",
      "\n",
      "dif is 0.09321776674443484 \n",
      "\n",
      "dif is 0.09116032957622665 \n",
      "\n",
      "dif is 0.08568658600115743 \n",
      "\n",
      "dif is 0.08884670015420669 \n",
      "\n",
      "dif is 0.08429338459758573 \n",
      "\n",
      "dif is 0.08745943950519963 \n",
      "\n",
      "dif is 0.08455223938768361 \n",
      "\n",
      "dif is 0.08441713332825183 \n",
      "\n",
      "dif is 0.08467415435424502 \n",
      "\n",
      "dif is 0.08224218027592779 \n",
      "\n",
      "dif is 0.08064961261282172 \n",
      "\n",
      "dif is 0.07811391043615114 \n",
      "\n",
      "dif is 0.0779307071681169 \n",
      "\n",
      "dif is 0.08005604466054674 \n",
      "\n",
      "dif is 0.07737586804563669 \n",
      "\n",
      "dif is 0.0743759765424531 \n",
      "\n",
      "dif is 0.07710976277652912 \n",
      "\n",
      "dif is 0.07065188894066757 \n",
      "\n",
      "dif is 0.07209275963625074 \n",
      "\n",
      "dif is 0.0712479159821328 \n",
      "\n",
      "dif is 0.07003251080338571 \n",
      "\n",
      "dif is 0.07515415788229954 \n",
      "\n",
      "dif is 0.06770472172802859 \n",
      "\n",
      "dif is 0.06783187336194141 \n",
      "\n",
      "dif is 0.07041586498406699 \n",
      "\n",
      "dif is 0.0689782106195188 \n",
      "\n",
      "dif is 0.06633060998179963 \n",
      "\n",
      "dif is 0.06250786497389445 \n",
      "\n",
      "dif is 0.06410401339308772 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 100 \n",
      "\n",
      "dif is 0.06261551558001612 \n",
      "\n",
      "dif is 0.06446608902058604 \n",
      "\n",
      "dif is 0.05998660884298479 \n",
      "\n",
      "dif is 0.06322537228666758 \n",
      "\n",
      "dif is 0.06209841274346392 \n",
      "\n",
      "dif is 0.06480567049277203 \n",
      "\n",
      "dif is 0.06026260962583821 \n",
      "\n",
      "dif is 0.06103056880428273 \n",
      "\n",
      "dif is 0.06411546379321549 \n",
      "\n",
      "dif is 0.06057776739920705 \n",
      "\n",
      "dif is 0.056606193378555075 \n",
      "\n",
      "dif is 0.05800642750770405 \n",
      "\n",
      "dif is 0.057770269110646776 \n",
      "\n",
      "dif is 0.06116352965725711 \n",
      "\n",
      "dif is 0.05755668893841431 \n",
      "\n",
      "dif is 0.05629700854484423 \n",
      "\n",
      "dif is 0.05582741034638362 \n",
      "\n",
      "dif is 0.05485504252337073 \n",
      "\n",
      "dif is 0.0551223854033454 \n",
      "\n",
      "dif is 0.05505106922765255 \n",
      "\n",
      "dif is 0.05579868301097106 \n",
      "\n",
      "dif is 0.05425927590523446 \n",
      "\n",
      "dif is 0.053823087015075134 \n",
      "\n",
      "dif is 0.05207350539436866 \n",
      "\n",
      "dif is 0.052573107385037426 \n",
      "\n",
      "dif is 0.05158633336012915 \n",
      "\n",
      "dif is 0.05184765395791349 \n",
      "\n",
      "dif is 0.050115770381587516 \n",
      "\n",
      "dif is 0.05143700717926607 \n",
      "\n",
      "dif is 0.04838216618576971 \n",
      "\n",
      "dif is 0.04885648467266026 \n",
      "\n",
      "dif is 0.048607421833052986 \n",
      "\n",
      "dif is 0.04781990467036341 \n",
      "\n",
      "dif is 0.04829797857716489 \n",
      "\n",
      "dif is 0.047082285006341186 \n",
      "\n",
      "dif is 0.04736362600815407 \n",
      "\n",
      "dif is 0.04871607287304715 \n",
      "\n",
      "dif is 0.045572166770452435 \n",
      "\n",
      "dif is 0.046608488796338415 \n",
      "\n",
      "dif is 0.04189037101573603 \n",
      "\n",
      "dif is 0.047328830867397316 \n",
      "\n",
      "dif is 0.045259478309959966 \n",
      "\n",
      "dif is 0.04404347951817042 \n",
      "\n",
      "dif is 0.045220952852744 \n",
      "\n",
      "dif is 0.0442024285689829 \n",
      "\n",
      "dif is 0.04570641789159342 \n",
      "\n",
      "dif is 0.043801981399086726 \n",
      "\n",
      "dif is 0.04424012897149793 \n",
      "\n",
      "dif is 0.0425834488937918 \n",
      "\n",
      "dif is 0.0439837350927892 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 150 \n",
      "\n",
      "dif is 0.042682267176138176 \n",
      "\n",
      "dif is 0.0452132229189881 \n",
      "\n",
      "dif is 0.04371633077088607 \n",
      "\n",
      "dif is 0.0400457856495274 \n",
      "\n",
      "dif is 0.043054962849926014 \n",
      "\n",
      "dif is 0.04414925845385856 \n",
      "\n",
      "dif is 0.04387626711593144 \n",
      "\n",
      "dif is 0.041529521717485085 \n",
      "\n",
      "dif is 0.03931450770259133 \n",
      "\n",
      "dif is 0.03978430222974519 \n",
      "\n",
      "dif is 0.04095566464673588 \n",
      "\n",
      "dif is 0.03927048713900893 \n",
      "\n",
      "dif is 0.038408744023937055 \n",
      "\n",
      "dif is 0.04079959544190961 \n",
      "\n",
      "dif is 0.037686796899982264 \n",
      "\n",
      "dif is 0.039133022024547735 \n",
      "\n",
      "dif is 0.04054526504634115 \n",
      "\n",
      "dif is 0.04047961980344799 \n",
      "\n",
      "dif is 0.04054466780867909 \n",
      "\n",
      "dif is 0.03997774920426691 \n",
      "\n",
      "dif is 0.037891646053861645 \n",
      "\n",
      "dif is 0.03699612890135897 \n",
      "\n",
      "dif is 0.03714380424754128 \n",
      "\n",
      "dif is 0.03504231762981094 \n",
      "\n",
      "dif is 0.038585675056168806 \n",
      "\n",
      "dif is 0.034992098166450235 \n",
      "\n",
      "dif is 0.037018439279247015 \n",
      "\n",
      "dif is 0.03803723533367907 \n",
      "\n",
      "dif is 0.0359232752630583 \n",
      "\n",
      "dif is 0.035991252685889284 \n",
      "\n",
      "dif is 0.035728840173325956 \n",
      "\n",
      "dif is 0.034633637889401284 \n",
      "\n",
      "dif is 0.034860554797646714 \n",
      "\n",
      "dif is 0.03395553876539458 \n",
      "\n",
      "dif is 0.03506689441302921 \n",
      "\n",
      "dif is 0.033835862502928846 \n",
      "\n",
      "dif is 0.03619689713944926 \n",
      "\n",
      "dif is 0.03409501978996391 \n",
      "\n",
      "dif is 0.03446384574762975 \n",
      "\n",
      "dif is 0.0354251134096112 \n",
      "\n",
      "dif is 0.03426029534179461 \n",
      "\n",
      "dif is 0.03658496119199478 \n",
      "\n",
      "dif is 0.03435396850388215 \n",
      "\n",
      "dif is 0.03329010393294554 \n",
      "\n",
      "dif is 0.03456857422942101 \n",
      "\n",
      "dif is 0.03378308293720834 \n",
      "\n",
      "dif is 0.03324669951335148 \n",
      "\n",
      "dif is 0.0327168681108586 \n",
      "\n",
      "dif is 0.03219648382094566 \n",
      "\n",
      "dif is 0.03317874984857302 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 200 \n",
      "\n",
      "dif is 0.03135331287586277 \n",
      "\n",
      "dif is 0.03204967983370686 \n",
      "\n",
      "dif is 0.031188112768064425 \n",
      "\n",
      "dif is 0.031439746305122786 \n",
      "\n",
      "dif is 0.03072655388670765 \n",
      "\n",
      "dif is 0.0306888331421144 \n",
      "\n",
      "dif is 0.03142042879125311 \n",
      "\n",
      "dif is 0.0317661767800731 \n",
      "\n",
      "dif is 0.030324712015508577 \n",
      "\n",
      "dif is 0.030954179894337443 \n",
      "\n",
      "dif is 0.03118537499424392 \n",
      "\n",
      "dif is 0.03223903832698478 \n",
      "\n",
      "dif is 0.029153436937807222 \n",
      "\n",
      "dif is 0.031626744682914695 \n",
      "\n",
      "dif is 0.030243110284403497 \n",
      "\n",
      "dif is 0.030700567271428936 \n",
      "\n",
      "dif is 0.028974067265442015 \n",
      "\n",
      "dif is 0.029891750594234642 \n",
      "\n",
      "dif is 0.029609448768412466 \n",
      "\n",
      "dif is 0.02919996531117356 \n",
      "\n",
      "dif is 0.02870317758166382 \n",
      "\n",
      "dif is 0.029250036072045977 \n",
      "\n",
      "dif is 0.030446074112789445 \n",
      "\n",
      "dif is 0.03060284537704506 \n",
      "\n",
      "dif is 0.029984384605651622 \n",
      "\n",
      "dif is 0.030088390042167814 \n",
      "\n",
      "dif is 0.028667423278235852 \n",
      "\n",
      "dif is 0.0280414262382026 \n",
      "\n",
      "dif is 0.028805218956274282 \n",
      "\n",
      "dif is 0.02800841157426049 \n",
      "\n",
      "dif is 0.026755231980112183 \n",
      "\n",
      "dif is 0.026937675136247712 \n",
      "\n",
      "dif is 0.028402319609427684 \n",
      "\n",
      "dif is 0.027080456700942782 \n",
      "\n",
      "dif is 0.027795742394016488 \n",
      "\n",
      "dif is 0.027946276735657385 \n",
      "\n",
      "dif is 0.026531594155916206 \n",
      "\n",
      "dif is 0.027742304218512898 \n",
      "\n",
      "dif is 0.027038961287872726 \n",
      "\n",
      "dif is 0.02663452624499298 \n",
      "\n",
      "dif is 0.02679239185646052 \n",
      "\n",
      "dif is 0.026739577313342606 \n",
      "\n",
      "dif is 0.025917719665650638 \n",
      "\n",
      "dif is 0.026376248231632284 \n",
      "\n",
      "dif is 0.027350827949995004 \n",
      "\n",
      "dif is 0.0251421835816541 \n",
      "\n",
      "dif is 0.026179663127222367 \n",
      "\n",
      "dif is 0.026863628253454744 \n",
      "\n",
      "dif is 0.02525001977667285 \n",
      "\n",
      "dif is 0.02689377293834312 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 250 \n",
      "\n",
      "dif is 0.02583416130638551 \n",
      "\n",
      "dif is 0.025767985444145252 \n",
      "\n",
      "dif is 0.02612734288947259 \n",
      "\n",
      "dif is 0.026424735582601555 \n",
      "\n",
      "dif is 0.02393363272532244 \n",
      "\n",
      "dif is 0.024308987563015774 \n",
      "\n",
      "dif is 0.026088369762059922 \n",
      "\n",
      "dif is 0.02545085772170823 \n",
      "\n",
      "dif is 0.024383246411742835 \n",
      "\n",
      "dif is 0.02467117438610766 \n",
      "\n",
      "dif is 0.02440196468015804 \n",
      "\n",
      "dif is 0.025185705686263743 \n",
      "\n",
      "dif is 0.024163024205589027 \n",
      "\n",
      "dif is 0.023938721948246762 \n",
      "\n",
      "dif is 0.024834935364764804 \n",
      "\n",
      "dif is 0.02363504344153509 \n",
      "\n",
      "dif is 0.024081253165884806 \n",
      "\n",
      "dif is 0.02346777105543627 \n",
      "\n",
      "dif is 0.023213277752772894 \n",
      "\n",
      "dif is 0.024865666368603854 \n",
      "\n",
      "dif is 0.025055150334516125 \n",
      "\n",
      "dif is 0.024402918917187696 \n",
      "\n",
      "dif is 0.024330896837246773 \n",
      "\n",
      "dif is 0.022921775352580304 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 0.023681718047293136 \n",
      "\n",
      "dif is 0.022426384646559227 \n",
      "\n",
      "dif is 0.023818966171720145 \n",
      "\n",
      "dif is 0.02297046307450034 \n",
      "\n",
      "dif is 0.023778745689652277 \n",
      "\n",
      "dif is 0.022655431881396085 \n",
      "\n",
      "dif is 0.02374775037854777 \n",
      "\n",
      "dif is 0.0243890671564934 \n",
      "\n",
      "dif is 0.023281597179218187 \n",
      "\n",
      "dif is 0.023270223798417946 \n",
      "\n",
      "dif is 0.02341545000616697 \n",
      "\n",
      "dif is 0.02325490860949931 \n",
      "\n",
      "dif is 0.021873608348011102 \n",
      "\n",
      "dif is 0.022662350566789485 \n",
      "\n",
      "dif is 0.02350958543662739 \n",
      "\n",
      "dif is 0.022695415419817004 \n",
      "\n",
      "dif is 0.023010640679894397 \n",
      "\n",
      "dif is 0.022046079840381465 \n",
      "\n",
      "dif is 0.023191939612313212 \n",
      "\n",
      "dif is 0.022147371050219564 \n",
      "\n",
      "dif is 0.022560358727873337 \n",
      "\n",
      "dif is 0.02181389673229712 \n",
      "\n",
      "dif is 0.02245706700630412 \n",
      "\n",
      "dif is 0.021911298859934335 \n",
      "\n",
      "dif is 0.021569160610807214 \n",
      "\n",
      "dif is 0.02252282639434814 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 300 \n",
      "\n",
      "dif is 0.02201570333901054 \n",
      "\n",
      "dif is 0.02153029380710425 \n",
      "\n",
      "dif is 0.021740667371673014 \n",
      "\n",
      "dif is 0.020962901400233255 \n",
      "\n",
      "dif is 0.02180085661741982 \n",
      "\n",
      "dif is 0.020285003000759888 \n",
      "\n",
      "dif is 0.021420529452082112 \n",
      "\n",
      "dif is 0.02112377268260323 \n",
      "\n",
      "dif is 0.021255096430166123 \n",
      "\n",
      "dif is 0.01990783060728542 \n",
      "\n",
      "dif is 0.021003160692073985 \n",
      "\n",
      "dif is 0.021270747315849062 \n",
      "\n",
      "dif is 0.020124893590025613 \n",
      "\n",
      "dif is 0.02046803985371876 \n",
      "\n",
      "dif is 0.01987048034737596 \n",
      "\n",
      "dif is 0.020683548923559066 \n",
      "\n",
      "dif is 0.019739749405113693 \n",
      "\n",
      "dif is 0.019771345800047968 \n",
      "\n",
      "dif is 0.019904975672196422 \n",
      "\n",
      "dif is 0.019358328971363224 \n",
      "\n",
      "dif is 0.01969904890496866 \n",
      "\n",
      "dif is 0.0207789720152803 \n",
      "\n",
      "dif is 0.02049625645631001 \n",
      "\n",
      "dif is 0.02013632878932538 \n",
      "\n",
      "dif is 0.020474089216858598 \n",
      "\n",
      "dif is 0.020143105994249986 \n",
      "\n",
      "dif is 0.020387978542717587 \n",
      "\n",
      "dif is 0.019694309384789863 \n",
      "\n",
      "dif is 0.02080120601915369 \n",
      "\n",
      "dif is 0.02020116461124002 \n",
      "\n",
      "dif is 0.019495796860038574 \n",
      "\n",
      "dif is 0.019296930902887914 \n",
      "\n",
      "dif is 0.019069313915878478 \n",
      "\n",
      "dif is 0.019427240391523895 \n",
      "\n",
      "dif is 0.019394880886167556 \n",
      "\n",
      "dif is 0.020378272339417844 \n",
      "\n",
      "dif is 0.018334940829340507 \n",
      "\n",
      "dif is 0.019451299324120448 \n",
      "\n",
      "dif is 0.01997131452435275 \n",
      "\n",
      "dif is 0.01882921172638912 \n",
      "\n",
      "dif is 0.018552107725851474 \n",
      "\n",
      "dif is 0.019156256582158568 \n",
      "\n",
      "dif is 0.01886640835432983 \n",
      "\n",
      "dif is 0.019620947908322094 \n",
      "\n",
      "dif is 0.019480490761511034 \n",
      "\n",
      "dif is 0.01886354179094905 \n",
      "\n",
      "dif is 0.019254661400855317 \n",
      "\n",
      "dif is 0.019557039794089456 \n",
      "\n",
      "dif is 0.018830191672990545 \n",
      "\n",
      "dif is 0.018828697270899093 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 350 \n",
      "\n",
      "dif is 0.018557882895671084 \n",
      "\n",
      "dif is 0.018598797692276275 \n",
      "\n",
      "dif is 0.01937458572600724 \n",
      "\n",
      "dif is 0.018677945652416022 \n",
      "\n",
      "dif is 0.018173664517994435 \n",
      "\n",
      "dif is 0.018321280451682954 \n",
      "\n",
      "dif is 0.018019951237198018 \n",
      "\n",
      "dif is 0.017850802025405583 \n",
      "\n",
      "dif is 0.0184786385102639 \n",
      "\n",
      "dif is 0.017998638911326744 \n",
      "\n",
      "dif is 0.018621164730558453 \n",
      "\n",
      "dif is 0.018524138752286833 \n",
      "\n",
      "dif is 0.018028792151494232 \n",
      "\n",
      "dif is 0.018204012765125618 \n",
      "\n",
      "dif is 0.017987321418517097 \n",
      "\n",
      "dif is 0.01774857219909195 \n",
      "\n",
      "dif is 0.0177590759667047 \n",
      "\n",
      "dif is 0.017502659638135517 \n",
      "\n",
      "dif is 0.01800240461103826 \n",
      "\n",
      "dif is 0.0172446850598788 \n",
      "\n",
      "dif is 0.01668205033593927 \n",
      "\n",
      "dif is 0.01692321604170257 \n",
      "\n",
      "dif is 0.01709058324167798 \n",
      "\n",
      "dif is 0.016915353599115057 \n",
      "\n",
      "dif is 0.017609018246751776 \n",
      "\n",
      "dif is 0.017464106543082083 \n",
      "\n",
      "dif is 0.01747391956383258 \n",
      "\n",
      "dif is 0.017388473592385437 \n",
      "\n",
      "dif is 0.01757267429867393 \n",
      "\n",
      "dif is 0.016567844107662374 \n",
      "\n",
      "dif is 0.016411233118652148 \n",
      "\n",
      "dif is 0.017147624293212777 \n",
      "\n",
      "dif is 0.01696943611153838 \n",
      "\n",
      "dif is 0.017113769745257995 \n",
      "\n",
      "dif is 0.01684909567575298 \n",
      "\n",
      "dif is 0.01646860830722833 \n",
      "\n",
      "dif is 0.016704899000570785 \n",
      "\n",
      "dif is 0.017041247858904287 \n",
      "\n",
      "dif is 0.016352925868293754 \n",
      "\n",
      "dif is 0.01708792671241847 \n",
      "\n",
      "dif is 0.01675365478971924 \n",
      "\n",
      "dif is 0.016024153101636764 \n",
      "\n",
      "dif is 0.015791310618778052 \n",
      "\n",
      "dif is 0.01658868125207073 \n",
      "\n",
      "dif is 0.016769234364392534 \n",
      "\n",
      "dif is 0.015631542833498754 \n",
      "\n",
      "dif is 0.016244137865289034 \n",
      "\n",
      "dif is 0.016279473932644152 \n",
      "\n",
      "dif is 0.016421340816621693 \n",
      "\n",
      "dif is 0.015656506753195708 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 400 \n",
      "\n",
      "dif is 0.01587139641160296 \n",
      "\n",
      "dif is 0.015446103188273752 \n",
      "\n",
      "dif is 0.016225809202609225 \n",
      "\n",
      "dif is 0.0156598064640967 \n",
      "\n",
      "dif is 0.01724446467639354 \n",
      "\n",
      "dif is 0.016956898690285428 \n",
      "\n",
      "dif is 0.01565559566982674 \n",
      "\n",
      "dif is 0.01625630393499259 \n",
      "\n",
      "dif is 0.015857154353809744 \n",
      "\n",
      "dif is 0.015017752682354963 \n",
      "\n",
      "dif is 0.016181910189648787 \n",
      "\n",
      "dif is 0.016341351677145514 \n",
      "\n",
      "dif is 0.01539346905583028 \n",
      "\n",
      "dif is 0.014976792967601215 \n",
      "\n",
      "dif is 0.01537023358650016 \n",
      "\n",
      "dif is 0.015542656550936116 \n",
      "\n",
      "dif is 0.015733417198759647 \n",
      "\n",
      "dif is 0.015553124086593463 \n",
      "\n",
      "dif is 0.014899128075910022 \n",
      "\n",
      "dif is 0.01584124044579738 \n",
      "\n",
      "dif is 0.0165307948689551 \n",
      "\n",
      "dif is 0.015499964869892349 \n",
      "\n",
      "dif is 0.01533043640823211 \n",
      "\n",
      "dif is 0.01544311221314502 \n",
      "\n",
      "dif is 0.015588931453591475 \n",
      "\n",
      "dif is 0.015427414368659088 \n",
      "\n",
      "dif is 0.01560928684974783 \n",
      "\n",
      "dif is 0.015564951042974862 \n",
      "\n",
      "dif is 0.014911721913531425 \n",
      "\n",
      "dif is 0.01474688113199906 \n",
      "\n",
      "dif is 0.015457468852555497 \n",
      "\n",
      "dif is 0.015384465112897133 \n",
      "\n",
      "dif is 0.014672939844515544 \n",
      "\n",
      "dif is 0.015144048728208353 \n",
      "\n",
      "dif is 0.014830572765154653 \n",
      "\n",
      "dif is 0.015069825358760496 \n",
      "\n",
      "dif is 0.014606507112951847 \n",
      "\n",
      "dif is 0.01466910987800547 \n",
      "\n",
      "dif is 0.014680326387663672 \n",
      "\n",
      "dif is 0.01445524265681846 \n",
      "\n",
      "dif is 0.015209438048150703 \n",
      "\n",
      "dif is 0.015281606114713131 \n",
      "\n",
      "dif is 0.014319840905436719 \n",
      "\n",
      "dif is 0.014519496513588599 \n",
      "\n",
      "dif is 0.014522641809966924 \n",
      "\n",
      "dif is 0.014105592236623506 \n",
      "\n",
      "dif is 0.014488877732898543 \n",
      "\n",
      "dif is 0.014424456716305427 \n",
      "\n",
      "dif is 0.014123492049358798 \n",
      "\n",
      "dif is 0.014962802029724725 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 450 \n",
      "\n",
      "dif is 0.014650649566859863 \n",
      "\n",
      "dif is 0.014703758487460318 \n",
      "\n",
      "dif is 0.01475351826554486 \n",
      "\n",
      "dif is 0.013690321825375143 \n",
      "\n",
      "dif is 0.014375646527200994 \n",
      "\n",
      "dif is 0.014035872241644352 \n",
      "\n",
      "dif is 0.014255120532853617 \n",
      "\n",
      "dif is 0.014790724636516062 \n",
      "\n",
      "dif is 0.013919913321352126 \n",
      "\n",
      "dif is 0.013660309083141156 \n",
      "\n",
      "dif is 0.014689519520701291 \n",
      "\n",
      "dif is 0.013676655138958556 \n",
      "\n",
      "dif is 0.013638979778080228 \n",
      "\n",
      "dif is 0.013664586774605393 \n",
      "\n",
      "dif is 0.014251359607839964 \n",
      "\n",
      "dif is 0.014750715355033599 \n",
      "\n",
      "dif is 0.01392788191709669 \n",
      "\n",
      "dif is 0.014120824987033286 \n",
      "\n",
      "dif is 0.013311664452099402 \n",
      "\n",
      "dif is 0.013701215050545574 \n",
      "\n",
      "dif is 0.013720237926727167 \n",
      "\n",
      "dif is 0.013269097906553543 \n",
      "\n",
      "dif is 0.014240906530767303 \n",
      "\n",
      "dif is 0.014558252594935643 \n",
      "\n",
      "dif is 0.01358991406959096 \n",
      "\n",
      "dif is 0.01389260796531849 \n",
      "\n",
      "dif is 0.013239197641963112 \n",
      "\n",
      "dif is 0.013845333629025245 \n",
      "\n",
      "dif is 0.013762168328585627 \n",
      "\n",
      "dif is 0.013277992719534384 \n",
      "\n",
      "dif is 0.01385974167390845 \n",
      "\n",
      "dif is 0.012362651119668357 \n",
      "\n",
      "dif is 0.013707620747341802 \n",
      "\n",
      "dif is 0.013930865098972258 \n",
      "\n",
      "dif is 0.012755755920738934 \n",
      "\n",
      "dif is 0.013431346893944793 \n",
      "\n",
      "dif is 0.013158874802781412 \n",
      "\n",
      "dif is 0.013023932172622552 \n",
      "\n",
      "dif is 0.013122929172701327 \n",
      "\n",
      "dif is 0.012532801505933033 \n",
      "\n",
      "dif is 0.013512986056363004 \n",
      "\n",
      "dif is 0.013382554288261031 \n",
      "\n",
      "dif is 0.013009069428401361 \n",
      "\n",
      "dif is 0.013343017656702566 \n",
      "\n",
      "dif is 0.01310117765197603 \n",
      "\n",
      "dif is 0.01304542671238997 \n",
      "\n",
      "dif is 0.012959163802600501 \n",
      "\n",
      "dif is 0.012605174029649852 \n",
      "\n",
      "dif is 0.013537683902944311 \n",
      "\n",
      "dif is 0.013174068726817794 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 500 \n",
      "\n",
      "dif is 0.01260264326794758 \n",
      "\n",
      "dif is 0.013373607979160847 \n",
      "\n",
      "dif is 0.012925471778512216 \n",
      "\n",
      "dif is 0.013403720357676133 \n",
      "\n",
      "dif is 0.013481687470185478 \n",
      "\n",
      "dif is 0.013089496412497662 \n",
      "\n",
      "dif is 0.012264659208503984 \n",
      "\n",
      "dif is 0.01280396940968123 \n",
      "\n",
      "dif is 0.013118922519005368 \n",
      "\n",
      "dif is 0.012732016895516163 \n",
      "\n",
      "dif is 0.012269518005275995 \n",
      "\n",
      "dif is 0.012931764144088365 \n",
      "\n",
      "dif is 0.012697897000220892 \n",
      "\n",
      "dif is 0.012490880194157148 \n",
      "\n",
      "dif is 0.013152407869765936 \n",
      "\n",
      "dif is 0.013009864328887047 \n",
      "\n",
      "dif is 0.012722869726879062 \n",
      "\n",
      "dif is 0.012083530349980432 \n",
      "\n",
      "dif is 0.013188488971499912 \n",
      "\n",
      "dif is 0.012397632727280396 \n",
      "\n",
      "dif is 0.012771908932531226 \n",
      "\n",
      "dif is 0.012281532777978615 \n",
      "\n",
      "dif is 0.012382699753459665 \n",
      "\n",
      "dif is 0.011613015255970275 \n",
      "\n",
      "dif is 0.012019862657353175 \n",
      "\n",
      "dif is 0.01225875684428068 \n",
      "\n",
      "dif is 0.0126182176116745 \n",
      "\n",
      "dif is 0.01202512012530863 \n",
      "\n",
      "dif is 0.011815965351873584 \n",
      "\n",
      "dif is 0.012003445217850383 \n",
      "\n",
      "dif is 0.012348911952518598 \n",
      "\n",
      "dif is 0.012146114288449375 \n",
      "\n",
      "dif is 0.011421037603907726 \n",
      "\n",
      "dif is 0.011525653034163598 \n",
      "\n",
      "dif is 0.011985315987206637 \n",
      "\n",
      "dif is 0.01197731128773409 \n",
      "\n",
      "dif is 0.012677008144104863 \n",
      "\n",
      "dif is 0.011684804044348273 \n",
      "\n",
      "dif is 0.011828491854298665 \n",
      "\n",
      "dif is 0.011993939600114524 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 0.01193366730661319 \n",
      "\n",
      "dif is 0.012057267729588613 \n",
      "\n",
      "dif is 0.012007958978953293 \n",
      "\n",
      "dif is 0.012010016846908 \n",
      "\n",
      "dif is 0.01278181759348681 \n",
      "\n",
      "dif is 0.011332240446553279 \n",
      "\n",
      "dif is 0.012067519734205147 \n",
      "\n",
      "dif is 0.012317277911723637 \n",
      "\n",
      "dif is 0.011319815180858736 \n",
      "\n",
      "dif is 0.011804437979910636 \n",
      "\n",
      "Burn-in sampling, finished full round thru SNP loci: 550 \n",
      "\n",
      "dif is 0.011944807991391557 \n",
      "\n",
      "dif is 0.011500089176719864 \n",
      "\n",
      "dif is 0.011251723353912668 \n",
      "\n",
      "dif is 0.011032914469402839 \n",
      "\n",
      "dif is 0.011565138950909038 \n",
      "\n",
      "dif is 0.011584472441211283 \n",
      "\n",
      "dif is 0.011650419148028836 \n",
      "\n",
      "dif is 0.0118401915828492 \n",
      "\n",
      "dif is 0.012063877004068486 \n",
      "\n",
      "dif is 0.011015585429401098 \n",
      "\n",
      "dif is 0.011120506708787167 \n",
      "\n",
      "dif is 0.011787781789737814 \n",
      "\n",
      "dif is 0.011020230085783478 \n",
      "\n",
      "dif is 0.011439012168974227 \n",
      "\n",
      "dif is 0.011042428150757993 \n",
      "\n",
      "dif is 0.01148537885541956 \n",
      "\n",
      "dif is 0.012008551543135368 \n",
      "\n",
      "dif is 0.011376640527464808 \n",
      "\n",
      "dif is 0.01164186915607843 \n",
      "\n",
      "dif is 0.010710687400476421 \n",
      "\n",
      "dif is 0.011017354807321302 \n",
      "\n",
      "dif is 0.011334771520391853 \n",
      "\n",
      "dif is 0.011356468194145063 \n",
      "\n",
      "dif is 0.011842923328281687 \n",
      "\n",
      "dif is 0.0118539115464342 \n",
      "\n",
      "dif is 0.0118294856158613 \n",
      "\n",
      "dif is 0.01081819421694731 \n",
      "\n",
      "dif is 0.011208129407732479 \n",
      "\n",
      "dif is 0.01120558271503563 \n",
      "\n",
      "dif is 0.011466976364475589 \n",
      "\n",
      "dif is 0.011302947595680356 \n",
      "\n",
      "dif is 0.011351546132327894 \n",
      "\n",
      "dif is 0.011608107662760648 \n",
      "\n",
      "dif is 0.009976763451980968 \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "706.4467251300812"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of burn-in rounds of Gibbs sampling: 583\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()  ## used to record time in system\n",
    "\n",
    "## outer loop: checking dif in gene sampling probabilities between previous two rounds\n",
    "##                and checking number of rounds \n",
    "while ((dif > stopping_threshold) and (burnin_rounds < max_burnin_rounds)):\n",
    "    ## run BEFORE iterating through genes for a round. \n",
    "    ## store count of genes in Gibbs samples for this round\n",
    "    burnin_gene_counts['round_ct'] = 0\n",
    "    \n",
    "    ## inner loop (the round): for each SNP loci...\n",
    "    for i in SNP_loci:  \n",
    "        ## note that if there is only one candidate gene for the SNP locus, \n",
    "        ##      the chunk doesn't run and the locus is SKIPPED (doesn't go thru sampling process)\n",
    "        if i not in SNPs_one_candidate:\n",
    "            ## grab Bayesian factor surrogate for each candidate gene in the current SNP locus i \n",
    "            currentloci_genes = candidate_genes[candidate_genes['SNP'] == i].filter( \\\n",
    "                                    items = ['alias_symbol', 'transformed_Fisherprod'])\n",
    "\n",
    "            ## look at the genes picked in other loci (not i)   \n",
    "            otherloci_genes = current_geneset[current_geneset['SNP'] != i]\n",
    "            ## take out genes that are also in SNP locus i \n",
    "            if flag_exclude_sharedrowcols is True:  \n",
    "                otherloci_genes = otherloci_genes[~ otherloci_genes['alias_symbol'].isin(\\\n",
    "                                                        currentloci_genes['alias_symbol'])]\n",
    "                ## if this takes out all of the genes, this would cause issues downstream. \n",
    "                ##      I haven't had this flag go off before. Perhaps can be dealt with by drawing a \n",
    "                ##      new current_geneset sample (like what is done at the beginning). \n",
    "                if len(otherloci_genes) == 0:\n",
    "                    raise ValueError(\\\n",
    "                        'Error: too much overlap between the gene set {genelist} and the SNP locus {SNP} led to issues calculating the network factor'.format(\\\n",
    "                        genelist = current_geneset['alias_symbol'].to_list(), SNP = i)) \n",
    "\n",
    "            ## retrieve network info: rows are candidate genes from SNP locus i, \n",
    "            ##                        columns are genes from other SNP loci (not i)\n",
    "            networkinfo = network_matrix.loc[currentloci_genes['alias_symbol'].to_list(), \\\n",
    "                                             otherloci_genes['alias_symbol'].to_list()]\n",
    "            network_factors = networkinfo.sum(axis = 'columns')  ## this is a pandas series \n",
    "            ## it is okay to do this because the series has the same row order as currentloci_genes\n",
    "            currentloci_genes['network_factor'] = network_factors.to_list()\n",
    "\n",
    "            ## calculate probability weighting of candidate genes using network factor and surrogate Bayesian factor     \n",
    "            currentloci_genes['probweights'] = \\\n",
    "                currentloci_genes['transformed_Fisherprod'] * currentloci_genes['network_factor']\n",
    "            ## if all the calculated weights are zero, set uniform prob dist \n",
    "            if currentloci_genes['probweights'].sum() == 0:\n",
    "                currentloci_genes['probweights'] = 1\n",
    "\n",
    "            ## choose loci's new state using the probability dist \n",
    "            newstate = currentloci_genes.sample(n = 1, axis = 0, weights = 'probweights')['alias_symbol'].to_list()\n",
    "            ## in Gibbs sample geneset, find row for SNP locus i, replace the entry with the new state\n",
    "            current_geneset.loc[i, 'alias_symbol'] = newstate    \n",
    "\n",
    "            ## record the current_geneset (Gibbs sample) in the round_ct column \n",
    "            ## using a for loop here with itertuples is ~30x slower \n",
    "            current_geneset['key'] = current_geneset['SNP'] + \",\" + current_geneset['alias_symbol']            \n",
    "            burnin_gene_counts.loc[burnin_gene_counts['key'].isin(current_geneset['key']), 'round_ct'] += 1\n",
    "            \n",
    "            ## no garbage collection here because it increases time for a loop by ~7x\n",
    "        ## end the if statement for only Gibbs sampling on SNP loci with > 1 candidate gene \n",
    "    ## end for loop of iterating thru each SNP loci    \n",
    "\n",
    "    ## a round is finished so increment this for stopping criteria \n",
    "    burnin_rounds += 1\n",
    "    if ((burnin_rounds % 50) == 0):\n",
    "        print(\"Burn-in sampling, finished full round thru SNP loci:\", burnin_rounds, \"\\n\")\n",
    "\n",
    "    ## add together the previous counts and new round_ct -> get new sampling counts\n",
    "    burnin_gene_counts['round_ct'] += burnin_gene_counts['counts'] \n",
    "\n",
    "    ## starting with the second round, calculate dif between sampling probabilities \n",
    "    ##    between previous round and current round\n",
    "    if burnin_rounds > 1:\n",
    "        ## sampling frequencies from previous round\n",
    "        burnin_gene_counts['freq_previous'] = burnin_gene_counts['counts'] / ((burnin_rounds - 1) * SNPs_per_round)\n",
    "        ## sampling frequencies from current round\n",
    "        burnin_gene_counts['freq_now'] = burnin_gene_counts['round_ct'] / (burnin_rounds * SNPs_per_round)\n",
    "        ## calculate sum of squared differences, then square-root this sum. \n",
    "        dif = math.sqrt(pd.Series.sum((burnin_gene_counts['freq_now'] - burnin_gene_counts['freq_previous']) ** 2))\n",
    "        print(\"dif is\", dif, \"\\n\")\n",
    "\n",
    "    ## replace old counts with current counts. \n",
    "    burnin_gene_counts['counts'] = burnin_gene_counts['round_ct']\n",
    "    ## drop freq columns if they're there\n",
    "    burnin_gene_counts = burnin_gene_counts[['SNP', 'alias_symbol', 'counts', 'key']]  \n",
    "\n",
    "## print the time it took to run\n",
    "t1 = time.time()\n",
    "burnin_time = t1 - t0\n",
    "burnin_time\n",
    "\n",
    "print(\"Number of burn-in rounds of Gibbs sampling:\", burnin_rounds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gibbs sampling: post-burn-in\n",
    "\n",
    "Keep the current_geneset from the last burn-in round. Start over the counting of genes sampled - these proportions should now approximate the posterior probability. \n",
    "\n",
    "**same changes and thoughts noted above for burn-in step**\n",
    "\n",
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>counts</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>0</td>\n",
       "      <td>rs115329265,HIST1H2BL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2AJ</td>\n",
       "      <td>0</td>\n",
       "      <td>rs115329265,HIST1H2AJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BM</td>\n",
       "      <td>0</td>\n",
       "      <td>rs115329265,HIST1H2BM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H2BN</td>\n",
       "      <td>0</td>\n",
       "      <td>rs115329265,HIST1H2BN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115329265</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>0</td>\n",
       "      <td>rs115329265,HIST1H1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>TAF7</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_140143664_I,TAF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1633</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>DIAPH1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_140143664_I,DIAPH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>HDAC3</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_140143664_I,HDAC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>ARAP3</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_140143664_I,ARAP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636</th>\n",
       "      <td>chr5_140143664_I</td>\n",
       "      <td>RELL2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_140143664_I,RELL2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1637 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNP alias_symbol  counts                      key\n",
       "0          rs115329265    HIST1H2BL       0    rs115329265,HIST1H2BL\n",
       "1          rs115329265    HIST1H2AJ       0    rs115329265,HIST1H2AJ\n",
       "2          rs115329265    HIST1H2BM       0    rs115329265,HIST1H2BM\n",
       "3          rs115329265    HIST1H2BN       0    rs115329265,HIST1H2BN\n",
       "4          rs115329265     HIST1H1B       0     rs115329265,HIST1H1B\n",
       "...                ...          ...     ...                      ...\n",
       "1632  chr5_140143664_I         TAF7       0    chr5_140143664_I,TAF7\n",
       "1633  chr5_140143664_I       DIAPH1       0  chr5_140143664_I,DIAPH1\n",
       "1634  chr5_140143664_I        HDAC3       0   chr5_140143664_I,HDAC3\n",
       "1635  chr5_140143664_I        ARAP3       0   chr5_140143664_I,ARAP3\n",
       "1636  chr5_140143664_I        RELL2       0   chr5_140143664_I,RELL2\n",
       "\n",
       "[1637 rows x 4 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## keep these variables: current_geneset from last burn-in round, \n",
    "##              SNP_loci, SNPs_one_candidate, SNPs_per_round\n",
    "\n",
    "## round = a set of sampling thru all loci\n",
    "max_final_rounds = 3000    ### maximum sampling round allowed at post burn in step\n",
    "\n",
    "## reset variables\n",
    "dif = stopping_threshold + 1  ## used for stopping criterion\n",
    "final_rounds = 0  ## counter for number of rounds thru all SNP loci\n",
    "\n",
    "## used to record counts of SNP-gene pairs from all previous rounds\n",
    "final_gene_counts = burnin_gene_counts.copy()\n",
    "## set up counter for candidate genes\n",
    "final_gene_counts['counts'] = 0\n",
    "\n",
    "final_gene_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Post-Burn-in Loop\n",
    "\n",
    "Starting with the second round of Gibbs sampling, it prints difference from previous round. \n",
    "\n",
    "It took me 766 seconds (12.8 minutes) to do 610 rounds to reach a difference below threshold. \n",
    "\n",
    "When I ran this for a second time for two post-burn-in loops (running both the setup code chunk and the following code chunk), the second time took me 807 seconds (13.45 min) to do 638 rounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 4.104327635875756 \n",
      "\n",
      "dif is 2.6409136989475455 \n",
      "\n",
      "dif is 1.745411294409637 \n",
      "\n",
      "dif is 1.4861547746167838 \n",
      "\n",
      "dif is 1.1230761730199248 \n",
      "\n",
      "dif is 0.9758704043080402 \n",
      "\n",
      "dif is 0.8093166070814872 \n",
      "\n",
      "dif is 0.7542798865491305 \n",
      "\n",
      "dif is 0.690340621772406 \n",
      "\n",
      "dif is 0.5858124893024745 \n",
      "\n",
      "dif is 0.5413065636763034 \n",
      "\n",
      "dif is 0.48604819561765816 \n",
      "\n",
      "dif is 0.46951089898835446 \n",
      "\n",
      "dif is 0.44428858394265713 \n",
      "\n",
      "dif is 0.4183222020082175 \n",
      "\n",
      "dif is 0.38438547894377495 \n",
      "\n",
      "dif is 0.3794956567020391 \n",
      "\n",
      "dif is 0.34581080114575313 \n",
      "\n",
      "dif is 0.32258005735979145 \n",
      "\n",
      "dif is 0.3265351924849056 \n",
      "\n",
      "dif is 0.2894390330395978 \n",
      "\n",
      "dif is 0.28518418205602625 \n",
      "\n",
      "dif is 0.26417674803458635 \n",
      "\n",
      "dif is 0.2606288608094724 \n",
      "\n",
      "dif is 0.25289490245631446 \n",
      "\n",
      "dif is 0.24789586206143413 \n",
      "\n",
      "dif is 0.23371959166274106 \n",
      "\n",
      "dif is 0.24000409383222895 \n",
      "\n",
      "dif is 0.22330144712538105 \n",
      "\n",
      "dif is 0.2102238258457328 \n",
      "\n",
      "dif is 0.20371332895842756 \n",
      "\n",
      "dif is 0.20452520863533025 \n",
      "\n",
      "dif is 0.19124360036902868 \n",
      "\n",
      "dif is 0.19145775965701478 \n",
      "\n",
      "dif is 0.18021643489614703 \n",
      "\n",
      "dif is 0.17423330470640566 \n",
      "\n",
      "dif is 0.16750099446513805 \n",
      "\n",
      "dif is 0.17059077490229843 \n",
      "\n",
      "dif is 0.15769969575736525 \n",
      "\n",
      "dif is 0.17101297366626209 \n",
      "\n",
      "dif is 0.15737592189543997 \n",
      "\n",
      "dif is 0.1602218618223215 \n",
      "\n",
      "dif is 0.14764842278132315 \n",
      "\n",
      "dif is 0.14355176388923982 \n",
      "\n",
      "dif is 0.14761744937831853 \n",
      "\n",
      "dif is 0.13986445910659664 \n",
      "\n",
      "dif is 0.13723451117497995 \n",
      "\n",
      "dif is 0.12730343555982737 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 50 \n",
      "\n",
      "dif is 0.13150505960680137 \n",
      "\n",
      "dif is 0.12710847255150473 \n",
      "\n",
      "dif is 0.12468137708864584 \n",
      "\n",
      "dif is 0.11741797228505546 \n",
      "\n",
      "dif is 0.11396570687804343 \n",
      "\n",
      "dif is 0.12410562910803874 \n",
      "\n",
      "dif is 0.1166428184603073 \n",
      "\n",
      "dif is 0.11515266065057589 \n",
      "\n",
      "dif is 0.11055686211788987 \n",
      "\n",
      "dif is 0.11331476115369274 \n",
      "\n",
      "dif is 0.10667237483987724 \n",
      "\n",
      "dif is 0.10593261689978085 \n",
      "\n",
      "dif is 0.10560295123498877 \n",
      "\n",
      "dif is 0.10917958349469713 \n",
      "\n",
      "dif is 0.10017221290775988 \n",
      "\n",
      "dif is 0.09302556227583346 \n",
      "\n",
      "dif is 0.10119846359821004 \n",
      "\n",
      "dif is 0.09675496098894576 \n",
      "\n",
      "dif is 0.09614139844456446 \n",
      "\n",
      "dif is 0.09714228605064297 \n",
      "\n",
      "dif is 0.08778321139944745 \n",
      "\n",
      "dif is 0.08972880714621784 \n",
      "\n",
      "dif is 0.09272971363296899 \n",
      "\n",
      "dif is 0.09466087379475835 \n",
      "\n",
      "dif is 0.08879099190984266 \n",
      "\n",
      "dif is 0.0876774209121806 \n",
      "\n",
      "dif is 0.08312954513293282 \n",
      "\n",
      "dif is 0.08471257487721158 \n",
      "\n",
      "dif is 0.08408445098538227 \n",
      "\n",
      "dif is 0.08230009962940817 \n",
      "\n",
      "dif is 0.08286254674160627 \n",
      "\n",
      "dif is 0.08036995579168929 \n",
      "\n",
      "dif is 0.07591192211472773 \n",
      "\n",
      "dif is 0.07392711781741294 \n",
      "\n",
      "dif is 0.07505242658293175 \n",
      "\n",
      "dif is 0.07670203593945495 \n",
      "\n",
      "dif is 0.07292576667800572 \n",
      "\n",
      "dif is 0.07564361346033048 \n",
      "\n",
      "dif is 0.0757026157921851 \n",
      "\n",
      "dif is 0.07302408005755101 \n",
      "\n",
      "dif is 0.07231627188700021 \n",
      "\n",
      "dif is 0.07268397719841606 \n",
      "\n",
      "dif is 0.07256072102634133 \n",
      "\n",
      "dif is 0.06828706283143361 \n",
      "\n",
      "dif is 0.06682402708997985 \n",
      "\n",
      "dif is 0.06966269245359773 \n",
      "\n",
      "dif is 0.06899365319522872 \n",
      "\n",
      "dif is 0.06539025294351994 \n",
      "\n",
      "dif is 0.0672249697708888 \n",
      "\n",
      "dif is 0.06110845645447173 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 100 \n",
      "\n",
      "dif is 0.06555247818738837 \n",
      "\n",
      "dif is 0.06416145567888301 \n",
      "\n",
      "dif is 0.060922989884013665 \n",
      "\n",
      "dif is 0.06335324118588992 \n",
      "\n",
      "dif is 0.05883471109861414 \n",
      "\n",
      "dif is 0.06286223344385171 \n",
      "\n",
      "dif is 0.05541808924209958 \n",
      "\n",
      "dif is 0.05702738812046099 \n",
      "\n",
      "dif is 0.06034559591033548 \n",
      "\n",
      "dif is 0.05800950937248737 \n",
      "\n",
      "dif is 0.056677526280569636 \n",
      "\n",
      "dif is 0.05947996765261867 \n",
      "\n",
      "dif is 0.0602949022875475 \n",
      "\n",
      "dif is 0.05573841046928992 \n",
      "\n",
      "dif is 0.05517854892307924 \n",
      "\n",
      "dif is 0.05437818450129525 \n",
      "\n",
      "dif is 0.05686325626804551 \n",
      "\n",
      "dif is 0.05802995117733595 \n",
      "\n",
      "dif is 0.054531055579125506 \n",
      "\n",
      "dif is 0.05398959740324459 \n",
      "\n",
      "dif is 0.05500883193125165 \n",
      "\n",
      "dif is 0.05380800942731598 \n",
      "\n",
      "dif is 0.05358983907100375 \n",
      "\n",
      "dif is 0.055318265361946575 \n",
      "\n",
      "dif is 0.0535831508213401 \n",
      "\n",
      "dif is 0.05439886919962283 \n",
      "\n",
      "dif is 0.05294986068339336 \n",
      "\n",
      "dif is 0.05143286314326728 \n",
      "\n",
      "dif is 0.05216889220708795 \n",
      "\n",
      "dif is 0.0498950772783279 \n",
      "\n",
      "dif is 0.05034378897771801 \n",
      "\n",
      "dif is 0.05037236412766512 \n",
      "\n",
      "dif is 0.04635221402008956 \n",
      "\n",
      "dif is 0.049215565178180806 \n",
      "\n",
      "dif is 0.048157043675699895 \n",
      "\n",
      "dif is 0.04923815560651028 \n",
      "\n",
      "dif is 0.046686672737090915 \n",
      "\n",
      "dif is 0.04638893645119733 \n",
      "\n",
      "dif is 0.045538114094698746 \n",
      "\n",
      "dif is 0.045295280830241556 \n",
      "\n",
      "dif is 0.044850146224488165 \n",
      "\n",
      "dif is 0.04474481067274164 \n",
      "\n",
      "dif is 0.044895681036637866 \n",
      "\n",
      "dif is 0.04489220040000694 \n",
      "\n",
      "dif is 0.04688532663047312 \n",
      "\n",
      "dif is 0.045724948762743474 \n",
      "\n",
      "dif is 0.043718477354032115 \n",
      "\n",
      "dif is 0.04441189826230827 \n",
      "\n",
      "dif is 0.04351240431177022 \n",
      "\n",
      "dif is 0.043669514839562024 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 150 \n",
      "\n",
      "dif is 0.043897865458827935 \n",
      "\n",
      "dif is 0.041924216193700924 \n",
      "\n",
      "dif is 0.0432152533018999 \n",
      "\n",
      "dif is 0.041943763152834636 \n",
      "\n",
      "dif is 0.04349315183777031 \n",
      "\n",
      "dif is 0.04266495844817638 \n",
      "\n",
      "dif is 0.04175635431016639 \n",
      "\n",
      "dif is 0.04086313062344976 \n",
      "\n",
      "dif is 0.04005810042172188 \n",
      "\n",
      "dif is 0.039139853815588196 \n",
      "\n",
      "dif is 0.04130590200404472 \n",
      "\n",
      "dif is 0.0396039187488334 \n",
      "\n",
      "dif is 0.04104148058962814 \n",
      "\n",
      "dif is 0.039679412890645256 \n",
      "\n",
      "dif is 0.037419404140286355 \n",
      "\n",
      "dif is 0.03846181746476186 \n",
      "\n",
      "dif is 0.04001402420354331 \n",
      "\n",
      "dif is 0.037847348131682744 \n",
      "\n",
      "dif is 0.03773639834855368 \n",
      "\n",
      "dif is 0.036608976833609176 \n",
      "\n",
      "dif is 0.03857125134013915 \n",
      "\n",
      "dif is 0.037492427189059235 \n",
      "\n",
      "dif is 0.03631048501502151 \n",
      "\n",
      "dif is 0.03789423092816294 \n",
      "\n",
      "dif is 0.037098739344084926 \n",
      "\n",
      "dif is 0.03662936271697775 \n",
      "\n",
      "dif is 0.03686591744778053 \n",
      "\n",
      "dif is 0.034980680283423535 \n",
      "\n",
      "dif is 0.03766658774370462 \n",
      "\n",
      "dif is 0.035708727815197315 \n",
      "\n",
      "dif is 0.03548987416660381 \n",
      "\n",
      "dif is 0.035607158496345 \n",
      "\n",
      "dif is 0.03550494325763293 \n",
      "\n",
      "dif is 0.03578654710000087 \n",
      "\n",
      "dif is 0.03522192925362095 \n",
      "\n",
      "dif is 0.03595978825177373 \n",
      "\n",
      "dif is 0.03354662108998748 \n",
      "\n",
      "dif is 0.03379473435152025 \n",
      "\n",
      "dif is 0.03429255346275757 \n",
      "\n",
      "dif is 0.03593361431639125 \n",
      "\n",
      "dif is 0.03424951997299252 \n",
      "\n",
      "dif is 0.03237278522321379 \n",
      "\n",
      "dif is 0.03448756327270642 \n",
      "\n",
      "dif is 0.03369399542522169 \n",
      "\n",
      "dif is 0.033083171895836073 \n",
      "\n",
      "dif is 0.03502196632724563 \n",
      "\n",
      "dif is 0.034193845701369734 \n",
      "\n",
      "dif is 0.03284002604456872 \n",
      "\n",
      "dif is 0.03044875422434001 \n",
      "\n",
      "dif is 0.03245745468883152 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 200 \n",
      "\n",
      "dif is 0.032610911641854305 \n",
      "\n",
      "dif is 0.03117993796216279 \n",
      "\n",
      "dif is 0.03303739717484367 \n",
      "\n",
      "dif is 0.030787338866742852 \n",
      "\n",
      "dif is 0.031655456309767654 \n",
      "\n",
      "dif is 0.03159431459863109 \n",
      "\n",
      "dif is 0.031868102684808906 \n",
      "\n",
      "dif is 0.03133559532086143 \n",
      "\n",
      "dif is 0.030434070535955635 \n",
      "\n",
      "dif is 0.030833844124394504 \n",
      "\n",
      "dif is 0.030971299303708007 \n",
      "\n",
      "dif is 0.029817194904880582 \n",
      "\n",
      "dif is 0.03011444550711512 \n",
      "\n",
      "dif is 0.0311285142271806 \n",
      "\n",
      "dif is 0.030111905300763485 \n",
      "\n",
      "dif is 0.02929272548997765 \n",
      "\n",
      "dif is 0.030263310687939297 \n",
      "\n",
      "dif is 0.029722280043041085 \n",
      "\n",
      "dif is 0.02952305375508722 \n",
      "\n",
      "dif is 0.030815691999916933 \n",
      "\n",
      "dif is 0.029753080762268506 \n",
      "\n",
      "dif is 0.028905558831695374 \n",
      "\n",
      "dif is 0.028615669626581484 \n",
      "\n",
      "dif is 0.029239048221949154 \n",
      "\n",
      "dif is 0.02901852814604947 \n",
      "\n",
      "dif is 0.029246867886642436 \n",
      "\n",
      "dif is 0.028548816608050146 \n",
      "\n",
      "dif is 0.028939513778585304 \n",
      "\n",
      "dif is 0.028262761466703906 \n",
      "\n",
      "dif is 0.027412296164030774 \n",
      "\n",
      "dif is 0.02846386736798024 \n",
      "\n",
      "dif is 0.028356880558082605 \n",
      "\n",
      "dif is 0.027754152111856608 \n",
      "\n",
      "dif is 0.026425882553035938 \n",
      "\n",
      "dif is 0.025728845137983777 \n",
      "\n",
      "dif is 0.028154674309055078 \n",
      "\n",
      "dif is 0.028817468832304457 \n",
      "\n",
      "dif is 0.027405606219471083 \n",
      "\n",
      "dif is 0.026953955085562455 \n",
      "\n",
      "dif is 0.026363907657026033 \n",
      "\n",
      "dif is 0.026343715839647945 \n",
      "\n",
      "dif is 0.027032797659681435 \n",
      "\n",
      "dif is 0.02853406516532028 \n",
      "\n",
      "dif is 0.026766917695886086 \n",
      "\n",
      "dif is 0.026229438464031182 \n",
      "\n",
      "dif is 0.025508965624891793 \n",
      "\n",
      "dif is 0.025922519106354663 \n",
      "\n",
      "dif is 0.02641055196360123 \n",
      "\n",
      "dif is 0.025804322216716125 \n",
      "\n",
      "dif is 0.02588712918375716 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 250 \n",
      "\n",
      "dif is 0.024973016973445956 \n",
      "\n",
      "dif is 0.025385032591974773 \n",
      "\n",
      "dif is 0.02572861630995584 \n",
      "\n",
      "dif is 0.025915335384271255 \n",
      "\n",
      "dif is 0.025962522454630706 \n",
      "\n",
      "dif is 0.02633147772399635 \n",
      "\n",
      "dif is 0.025848710086588962 \n",
      "\n",
      "dif is 0.0258942345085736 \n",
      "\n",
      "dif is 0.025826235225239132 \n",
      "\n",
      "dif is 0.025946205581029076 \n",
      "\n",
      "dif is 0.024322508034648752 \n",
      "\n",
      "dif is 0.025587217573469507 \n",
      "\n",
      "dif is 0.025323596528205112 \n",
      "\n",
      "dif is 0.024601885108974625 \n",
      "\n",
      "dif is 0.024711325425925374 \n",
      "\n",
      "dif is 0.02422731984046195 \n",
      "\n",
      "dif is 0.02354746394577284 \n",
      "\n",
      "dif is 0.024748143820294194 \n",
      "\n",
      "dif is 0.02436450021031254 \n",
      "\n",
      "dif is 0.02443174301277825 \n",
      "\n",
      "dif is 0.023119542789364442 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 0.024383310043854828 \n",
      "\n",
      "dif is 0.024794494627562086 \n",
      "\n",
      "dif is 0.023087208326457016 \n",
      "\n",
      "dif is 0.0223790602800941 \n",
      "\n",
      "dif is 0.024123707561516328 \n",
      "\n",
      "dif is 0.023884925203878676 \n",
      "\n",
      "dif is 0.02437537120392872 \n",
      "\n",
      "dif is 0.02266806261168412 \n",
      "\n",
      "dif is 0.023130706446494485 \n",
      "\n",
      "dif is 0.022875561894457774 \n",
      "\n",
      "dif is 0.023808006929003814 \n",
      "\n",
      "dif is 0.0230941097131435 \n",
      "\n",
      "dif is 0.023007376966996264 \n",
      "\n",
      "dif is 0.022243353371990973 \n",
      "\n",
      "dif is 0.02337731624061857 \n",
      "\n",
      "dif is 0.02356692559625294 \n",
      "\n",
      "dif is 0.022884620244764155 \n",
      "\n",
      "dif is 0.022181411625549683 \n",
      "\n",
      "dif is 0.022362724281247255 \n",
      "\n",
      "dif is 0.02287508421746003 \n",
      "\n",
      "dif is 0.022615579869567207 \n",
      "\n",
      "dif is 0.022156210907674827 \n",
      "\n",
      "dif is 0.02211890666304245 \n",
      "\n",
      "dif is 0.021368173616553292 \n",
      "\n",
      "dif is 0.022162596197695332 \n",
      "\n",
      "dif is 0.02285919599843703 \n",
      "\n",
      "dif is 0.022046573407137626 \n",
      "\n",
      "dif is 0.021549248360725207 \n",
      "\n",
      "dif is 0.021066641665357422 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 300 \n",
      "\n",
      "dif is 0.022568976416755242 \n",
      "\n",
      "dif is 0.02213581543390596 \n",
      "\n",
      "dif is 0.021643025493974205 \n",
      "\n",
      "dif is 0.020669566128950607 \n",
      "\n",
      "dif is 0.021720674968134235 \n",
      "\n",
      "dif is 0.02120406962842216 \n",
      "\n",
      "dif is 0.022336906944190667 \n",
      "\n",
      "dif is 0.021062392133984482 \n",
      "\n",
      "dif is 0.02074743302261731 \n",
      "\n",
      "dif is 0.021648285588841354 \n",
      "\n",
      "dif is 0.021020964473279967 \n",
      "\n",
      "dif is 0.0212209582805383 \n",
      "\n",
      "dif is 0.02150931022207568 \n",
      "\n",
      "dif is 0.02136668015708584 \n",
      "\n",
      "dif is 0.020210474445384968 \n",
      "\n",
      "dif is 0.020204351350014388 \n",
      "\n",
      "dif is 0.01980435329634779 \n",
      "\n",
      "dif is 0.018733811700895747 \n",
      "\n",
      "dif is 0.02029635267066156 \n",
      "\n",
      "dif is 0.02024819514287073 \n",
      "\n",
      "dif is 0.02041192278051196 \n",
      "\n",
      "dif is 0.019544853224734968 \n",
      "\n",
      "dif is 0.01962057204535388 \n",
      "\n",
      "dif is 0.019209448163114203 \n",
      "\n",
      "dif is 0.019470401636831867 \n",
      "\n",
      "dif is 0.01958605538375987 \n",
      "\n",
      "dif is 0.01916542577167093 \n",
      "\n",
      "dif is 0.018555565587507553 \n",
      "\n",
      "dif is 0.02025087564578126 \n",
      "\n",
      "dif is 0.019568447122106817 \n",
      "\n",
      "dif is 0.019153687384110597 \n",
      "\n",
      "dif is 0.019104628767698806 \n",
      "\n",
      "dif is 0.01838455146712629 \n",
      "\n",
      "dif is 0.01947415915705188 \n",
      "\n",
      "dif is 0.019217200499518524 \n",
      "\n",
      "dif is 0.018888394789400874 \n",
      "\n",
      "dif is 0.018962385248275065 \n",
      "\n",
      "dif is 0.01880613848761959 \n",
      "\n",
      "dif is 0.018678799921371787 \n",
      "\n",
      "dif is 0.01913638556258264 \n",
      "\n",
      "dif is 0.01992393339816456 \n",
      "\n",
      "dif is 0.019303120485364644 \n",
      "\n",
      "dif is 0.018488887732899847 \n",
      "\n",
      "dif is 0.01897190920053946 \n",
      "\n",
      "dif is 0.019344076987987725 \n",
      "\n",
      "dif is 0.01785226883701421 \n",
      "\n",
      "dif is 0.01917365925404417 \n",
      "\n",
      "dif is 0.018356429927682417 \n",
      "\n",
      "dif is 0.018701738164642316 \n",
      "\n",
      "dif is 0.017724033212367696 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 350 \n",
      "\n",
      "dif is 0.017794758950897802 \n",
      "\n",
      "dif is 0.019252282599878254 \n",
      "\n",
      "dif is 0.0189673246855819 \n",
      "\n",
      "dif is 0.018899954955773898 \n",
      "\n",
      "dif is 0.01838193621172453 \n",
      "\n",
      "dif is 0.017881885255857582 \n",
      "\n",
      "dif is 0.01939575549387996 \n",
      "\n",
      "dif is 0.018533669363424283 \n",
      "\n",
      "dif is 0.01880501878695706 \n",
      "\n",
      "dif is 0.017561331806249594 \n",
      "\n",
      "dif is 0.016961902757513348 \n",
      "\n",
      "dif is 0.017363053306758677 \n",
      "\n",
      "dif is 0.01749701904825454 \n",
      "\n",
      "dif is 0.017642261719982545 \n",
      "\n",
      "dif is 0.017064656225277076 \n",
      "\n",
      "dif is 0.017627146883816586 \n",
      "\n",
      "dif is 0.01701228411499975 \n",
      "\n",
      "dif is 0.01681577361635302 \n",
      "\n",
      "dif is 0.017977670112164407 \n",
      "\n",
      "dif is 0.017652875892907472 \n",
      "\n",
      "dif is 0.01795236997764865 \n",
      "\n",
      "dif is 0.017336403688760146 \n",
      "\n",
      "dif is 0.017496925668687494 \n",
      "\n",
      "dif is 0.01706432136419069 \n",
      "\n",
      "dif is 0.017700210572250158 \n",
      "\n",
      "dif is 0.016852386591019397 \n",
      "\n",
      "dif is 0.017768839463973195 \n",
      "\n",
      "dif is 0.0177592793192455 \n",
      "\n",
      "dif is 0.016360865930343048 \n",
      "\n",
      "dif is 0.01735759733285134 \n",
      "\n",
      "dif is 0.01692213986727443 \n",
      "\n",
      "dif is 0.016901110471461662 \n",
      "\n",
      "dif is 0.01738618592294015 \n",
      "\n",
      "dif is 0.017477895482762816 \n",
      "\n",
      "dif is 0.017633288029252646 \n",
      "\n",
      "dif is 0.01695768449282781 \n",
      "\n",
      "dif is 0.016504113549029056 \n",
      "\n",
      "dif is 0.01684659320292645 \n",
      "\n",
      "dif is 0.016763659028227688 \n",
      "\n",
      "dif is 0.0172950087891184 \n",
      "\n",
      "dif is 0.01718585211955061 \n",
      "\n",
      "dif is 0.016063898050581928 \n",
      "\n",
      "dif is 0.01662642772875688 \n",
      "\n",
      "dif is 0.01704548826972849 \n",
      "\n",
      "dif is 0.01758662615234324 \n",
      "\n",
      "dif is 0.017037073296201733 \n",
      "\n",
      "dif is 0.016123376187901665 \n",
      "\n",
      "dif is 0.015883822190765125 \n",
      "\n",
      "dif is 0.016404611945535337 \n",
      "\n",
      "dif is 0.01600100890349919 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 400 \n",
      "\n",
      "dif is 0.016137555466273473 \n",
      "\n",
      "dif is 0.015940817347944832 \n",
      "\n",
      "dif is 0.015553841412396394 \n",
      "\n",
      "dif is 0.015848714530259916 \n",
      "\n",
      "dif is 0.016343149581821835 \n",
      "\n",
      "dif is 0.016675677631088626 \n",
      "\n",
      "dif is 0.01682551355691399 \n",
      "\n",
      "dif is 0.016337300266613248 \n",
      "\n",
      "dif is 0.01534399529821249 \n",
      "\n",
      "dif is 0.01567276779958916 \n",
      "\n",
      "dif is 0.016071121735177222 \n",
      "\n",
      "dif is 0.015732768795696413 \n",
      "\n",
      "dif is 0.015487263227863042 \n",
      "\n",
      "dif is 0.015590749136791187 \n",
      "\n",
      "dif is 0.014750741081689693 \n",
      "\n",
      "dif is 0.015661999568792932 \n",
      "\n",
      "dif is 0.015480565384313095 \n",
      "\n",
      "dif is 0.015499955528114208 \n",
      "\n",
      "dif is 0.014840145825245133 \n",
      "\n",
      "dif is 0.014954508123984083 \n",
      "\n",
      "dif is 0.01557884853211512 \n",
      "\n",
      "dif is 0.015549954481525166 \n",
      "\n",
      "dif is 0.016325358671714855 \n",
      "\n",
      "dif is 0.016002435037115282 \n",
      "\n",
      "dif is 0.01548182648628468 \n",
      "\n",
      "dif is 0.014765657495515625 \n",
      "\n",
      "dif is 0.015172660619788142 \n",
      "\n",
      "dif is 0.014665224049828706 \n",
      "\n",
      "dif is 0.01552289338788771 \n",
      "\n",
      "dif is 0.015900314745404615 \n",
      "\n",
      "dif is 0.015758797883038498 \n",
      "\n",
      "dif is 0.014375706809071577 \n",
      "\n",
      "dif is 0.015381233564501088 \n",
      "\n",
      "dif is 0.014726756699899925 \n",
      "\n",
      "dif is 0.015054503784491921 \n",
      "\n",
      "dif is 0.014876216768489948 \n",
      "\n",
      "dif is 0.01492720432245876 \n",
      "\n",
      "dif is 0.014738169614036406 \n",
      "\n",
      "dif is 0.014571829263733543 \n",
      "\n",
      "dif is 0.01468999904997374 \n",
      "\n",
      "dif is 0.015017690370957681 \n",
      "\n",
      "dif is 0.014972955328505769 \n",
      "\n",
      "dif is 0.014670302596584252 \n",
      "\n",
      "dif is 0.01510531067359697 \n",
      "\n",
      "dif is 0.014633188872988345 \n",
      "\n",
      "dif is 0.014245007142565325 \n",
      "\n",
      "dif is 0.013953877875902475 \n",
      "\n",
      "dif is 0.014495930992496079 \n",
      "\n",
      "dif is 0.01459418550347129 \n",
      "\n",
      "dif is 0.013714048211784487 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 450 \n",
      "\n",
      "dif is 0.014257544387860904 \n",
      "\n",
      "dif is 0.013547882028644438 \n",
      "\n",
      "dif is 0.014784115735102382 \n",
      "\n",
      "dif is 0.014754552894953062 \n",
      "\n",
      "dif is 0.013938068448562284 \n",
      "\n",
      "dif is 0.014213975682530411 \n",
      "\n",
      "dif is 0.013833624495017838 \n",
      "\n",
      "dif is 0.013970837307989079 \n",
      "\n",
      "dif is 0.01429540281652964 \n",
      "\n",
      "dif is 0.014528202224337388 \n",
      "\n",
      "dif is 0.014050570594046144 \n",
      "\n",
      "dif is 0.014114713255360602 \n",
      "\n",
      "dif is 0.014305966973346348 \n",
      "\n",
      "dif is 0.0138228056060426 \n",
      "\n",
      "dif is 0.014298856998454004 \n",
      "\n",
      "dif is 0.013109704416327203 \n",
      "\n",
      "dif is 0.014320653632355074 \n",
      "\n",
      "dif is 0.013634866802402393 \n",
      "\n",
      "dif is 0.014377539576050976 \n",
      "\n",
      "dif is 0.013377363206458945 \n",
      "\n",
      "dif is 0.014275146633256021 \n",
      "\n",
      "dif is 0.013865296820605014 \n",
      "\n",
      "dif is 0.014024859829864272 \n",
      "\n",
      "dif is 0.013725088135938422 \n",
      "\n",
      "dif is 0.013390233500030423 \n",
      "\n",
      "dif is 0.013399554139847917 \n",
      "\n",
      "dif is 0.013101159845130576 \n",
      "\n",
      "dif is 0.01399627274058053 \n",
      "\n",
      "dif is 0.013188592964600647 \n",
      "\n",
      "dif is 0.01396590978443933 \n",
      "\n",
      "dif is 0.013191406645880201 \n",
      "\n",
      "dif is 0.01351953485392662 \n",
      "\n",
      "dif is 0.013714348223115524 \n",
      "\n",
      "dif is 0.013043083694822636 \n",
      "\n",
      "dif is 0.013636012265089982 \n",
      "\n",
      "dif is 0.013194140611427076 \n",
      "\n",
      "dif is 0.013742254413289379 \n",
      "\n",
      "dif is 0.013407109973967907 \n",
      "\n",
      "dif is 0.013554602163778885 \n",
      "\n",
      "dif is 0.013199036815197945 \n",
      "\n",
      "dif is 0.01330634124219938 \n",
      "\n",
      "dif is 0.01284449985373993 \n",
      "\n",
      "dif is 0.013420330047856859 \n",
      "\n",
      "dif is 0.012658533081052793 \n",
      "\n",
      "dif is 0.012704914781252446 \n",
      "\n",
      "dif is 0.012644286612180502 \n",
      "\n",
      "dif is 0.0129202223285506 \n",
      "\n",
      "dif is 0.012833685502083244 \n",
      "\n",
      "dif is 0.012250251999830973 \n",
      "\n",
      "dif is 0.012802764453384517 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 500 \n",
      "\n",
      "dif is 0.013017420998947917 \n",
      "\n",
      "dif is 0.013637458569349021 \n",
      "\n",
      "dif is 0.012972015826550975 \n",
      "\n",
      "dif is 0.013330496795146328 \n",
      "\n",
      "dif is 0.01314342068927641 \n",
      "\n",
      "dif is 0.01285374464484274 \n",
      "\n",
      "dif is 0.012365096192886645 \n",
      "\n",
      "dif is 0.01235773892197039 \n",
      "\n",
      "dif is 0.012423278203290611 \n",
      "\n",
      "dif is 0.012846417152778282 \n",
      "\n",
      "dif is 0.013176869335490872 \n",
      "\n",
      "dif is 0.012443780370228573 \n",
      "\n",
      "dif is 0.012644694109316639 \n",
      "\n",
      "dif is 0.012185269025483668 \n",
      "\n",
      "dif is 0.012172956776389503 \n",
      "\n",
      "dif is 0.012052566462946319 \n",
      "\n",
      "dif is 0.012720081920672253 \n",
      "\n",
      "dif is 0.012518423138776172 \n",
      "\n",
      "dif is 0.012672625266053648 \n",
      "\n",
      "dif is 0.012295497688620741 \n",
      "\n",
      "dif is 0.01213793918693631 \n",
      "\n",
      "dif is 0.012149369465351457 \n",
      "\n",
      "dif is 0.012338981370351285 \n",
      "\n",
      "dif is 0.012924881941493974 \n",
      "\n",
      "dif is 0.012938183681793206 \n",
      "\n",
      "dif is 0.012450773646084748 \n",
      "\n",
      "dif is 0.01250891413230631 \n",
      "\n",
      "dif is 0.011406473693499099 \n",
      "\n",
      "dif is 0.012867771058486731 \n",
      "\n",
      "dif is 0.01280175766668911 \n",
      "\n",
      "dif is 0.01255749631938014 \n",
      "\n",
      "dif is 0.01230875834927454 \n",
      "\n",
      "dif is 0.011999399656490651 \n",
      "\n",
      "dif is 0.011788482684096775 \n",
      "\n",
      "dif is 0.011603217799466001 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dif is 0.012121833894403907 \n",
      "\n",
      "dif is 0.01237976726447999 \n",
      "\n",
      "dif is 0.011879016399236133 \n",
      "\n",
      "dif is 0.011759740198140025 \n",
      "\n",
      "dif is 0.01194430576827477 \n",
      "\n",
      "dif is 0.012297341602868475 \n",
      "\n",
      "dif is 0.012190901927987077 \n",
      "\n",
      "dif is 0.011483977289683912 \n",
      "\n",
      "dif is 0.012177829373136355 \n",
      "\n",
      "dif is 0.011915067137766962 \n",
      "\n",
      "dif is 0.011987546577340172 \n",
      "\n",
      "dif is 0.01205739416760665 \n",
      "\n",
      "dif is 0.011984386726753806 \n",
      "\n",
      "dif is 0.011719707867733168 \n",
      "\n",
      "dif is 0.011070173302404895 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 550 \n",
      "\n",
      "dif is 0.011531184429542573 \n",
      "\n",
      "dif is 0.011734118468568894 \n",
      "\n",
      "dif is 0.011692072479046314 \n",
      "\n",
      "dif is 0.011777311983586657 \n",
      "\n",
      "dif is 0.011849189296858258 \n",
      "\n",
      "dif is 0.011591228255443287 \n",
      "\n",
      "dif is 0.011128174142941018 \n",
      "\n",
      "dif is 0.011925084539346346 \n",
      "\n",
      "dif is 0.012123720300576261 \n",
      "\n",
      "dif is 0.010962567338860151 \n",
      "\n",
      "dif is 0.01179684385245157 \n",
      "\n",
      "dif is 0.011407802799371834 \n",
      "\n",
      "dif is 0.011035646638847398 \n",
      "\n",
      "dif is 0.011060412011670436 \n",
      "\n",
      "dif is 0.0116919522448791 \n",
      "\n",
      "dif is 0.011338700004516221 \n",
      "\n",
      "dif is 0.011219568120410518 \n",
      "\n",
      "dif is 0.01122126953354531 \n",
      "\n",
      "dif is 0.011259016933731256 \n",
      "\n",
      "dif is 0.011434572029144177 \n",
      "\n",
      "dif is 0.011378134253390619 \n",
      "\n",
      "dif is 0.01054415662951376 \n",
      "\n",
      "dif is 0.011242505878480778 \n",
      "\n",
      "dif is 0.011435681990479382 \n",
      "\n",
      "dif is 0.010862330818956297 \n",
      "\n",
      "dif is 0.011303289399610976 \n",
      "\n",
      "dif is 0.011708724390233254 \n",
      "\n",
      "dif is 0.01169517470182546 \n",
      "\n",
      "dif is 0.012091382707919046 \n",
      "\n",
      "dif is 0.011125841899604066 \n",
      "\n",
      "dif is 0.011289116973093477 \n",
      "\n",
      "dif is 0.010753037084381536 \n",
      "\n",
      "dif is 0.011124129652377326 \n",
      "\n",
      "dif is 0.011029069324929251 \n",
      "\n",
      "dif is 0.011125385239519038 \n",
      "\n",
      "dif is 0.011169237794409752 \n",
      "\n",
      "dif is 0.01072135825192225 \n",
      "\n",
      "dif is 0.01118992219269395 \n",
      "\n",
      "dif is 0.010709168184292503 \n",
      "\n",
      "dif is 0.010645020284258361 \n",
      "\n",
      "dif is 0.010652352848868807 \n",
      "\n",
      "dif is 0.010515202560037977 \n",
      "\n",
      "dif is 0.010829588233959096 \n",
      "\n",
      "dif is 0.011095546941413506 \n",
      "\n",
      "dif is 0.010663671828954013 \n",
      "\n",
      "dif is 0.010869114157994434 \n",
      "\n",
      "dif is 0.010507646066599789 \n",
      "\n",
      "dif is 0.010984256268902731 \n",
      "\n",
      "dif is 0.01138730924043399 \n",
      "\n",
      "dif is 0.011122865363160261 \n",
      "\n",
      "Final (post-burn-in) sampling, finished full round thru SNP loci: 600 \n",
      "\n",
      "dif is 0.01078532183231444 \n",
      "\n",
      "dif is 0.0106640443744231 \n",
      "\n",
      "dif is 0.01094183591785589 \n",
      "\n",
      "dif is 0.010596063271480627 \n",
      "\n",
      "dif is 0.01068015709128982 \n",
      "\n",
      "dif is 0.010946729293236583 \n",
      "\n",
      "dif is 0.011183077952872279 \n",
      "\n",
      "dif is 0.010812478652468253 \n",
      "\n",
      "dif is 0.011175262900822349 \n",
      "\n",
      "dif is 0.011058688290772712 \n",
      "\n",
      "dif is 0.010175946668805533 \n",
      "\n",
      "dif is 0.01080658813596062 \n",
      "\n",
      "dif is 0.010570539426157327 \n",
      "\n",
      "dif is 0.010520312444705738 \n",
      "\n",
      "dif is 0.010642498375672521 \n",
      "\n",
      "dif is 0.010354010985867013 \n",
      "\n",
      "dif is 0.010337782557669147 \n",
      "\n",
      "dif is 0.010284792147102807 \n",
      "\n",
      "dif is 0.010425994236157622 \n",
      "\n",
      "dif is 0.01015254610118618 \n",
      "\n",
      "dif is 0.01029749315668021 \n",
      "\n",
      "dif is 0.009978267761195385 \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "736.2830719947815"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of final (post-burn-in) rounds of Gibbs sampling: 621\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()  ## used to record time in system\n",
    "\n",
    "## outer loop: checking dif in gene sampling probabilities between previous two rounds\n",
    "##                and checking number of rounds \n",
    "while ((dif > stopping_threshold) and (final_rounds < max_final_rounds)):\n",
    "    ## run BEFORE iterating through genes for a round. \n",
    "    ## store count of genes in Gibbs samples for this round\n",
    "    final_gene_counts['round_ct'] = 0\n",
    "    \n",
    "    ## inner loop (the round): for each SNP loci...    \n",
    "    for i in SNP_loci: \n",
    "        ## note that if there is only one candidate gene for the SNP locus, \n",
    "        ##      the chunk doesn't run and the locus is SKIPPED (doesn't go thru sampling process)\n",
    "        if i not in SNPs_one_candidate:\n",
    "            ## grab Bayesian factor surrogate for each candidate gene in the current SNP locus i \n",
    "            currentloci_genes = candidate_genes[candidate_genes['SNP'] == i].filter( \\\n",
    "                                    items = ['alias_symbol', 'transformed_Fisherprod'])\n",
    "\n",
    "            ## look at the genes picked in other loci (not i)    \n",
    "            otherloci_genes = current_geneset[current_geneset['SNP'] != i]\n",
    "            ## take out genes that are also in SNP locus i \n",
    "            if flag_exclude_sharedrowcols is True:  \n",
    "                otherloci_genes = otherloci_genes[~ otherloci_genes['alias_symbol'].isin(\\\n",
    "                                                        currentloci_genes['alias_symbol'])]\n",
    "                ## if this takes out all of the genes, this would cause issues downstream. \n",
    "                ##      I haven't had this flag go off before. Perhaps can be dealt with by drawing a \n",
    "                ##      new current_geneset sample (like what is done at the beginning). \n",
    "                if len(otherloci_genes) == 0:\n",
    "                    raise ValueError(\\\n",
    "                        'Error: too much overlap between the gene set {genelist} and the SNP locus {SNP} led to issues calculating the network factor'.format(\\\n",
    "                        genelist = current_geneset['alias_symbol'].to_list(), SNP = i)) \n",
    "\n",
    "            ## retrieve network info: rows are candidate genes from SNP locus i, \n",
    "            ##                        columns are genes from other SNP loci (not i)\n",
    "            networkinfo = network_matrix.loc[currentloci_genes['alias_symbol'].to_list(), \\\n",
    "                                             otherloci_genes['alias_symbol'].to_list()]\n",
    "            network_factors = networkinfo.sum(axis = 'columns')  ## this is a pandas series \n",
    "            ## it is okay to do this because the series has the same row order as currentloci_genes\n",
    "            currentloci_genes['network_factor'] = network_factors.to_list()\n",
    "\n",
    "            ## calculate probability weighting of candidate genes using network factor and surrogate Bayesian factor     \n",
    "            currentloci_genes['probweights'] = currentloci_genes['transformed_Fisherprod'] * currentloci_genes['network_factor']\n",
    "            ## if all the calculated weights are zero, set uniform prob dist \n",
    "            if currentloci_genes['probweights'].sum() == 0:\n",
    "                currentloci_genes['probweights'] = 1\n",
    "\n",
    "            ## choose loci's new state using the probability dist \n",
    "            newstate = currentloci_genes.sample(n = 1, axis = 0, weights = 'probweights')['alias_symbol'].to_list()\n",
    "            ## in Gibbs sample geneset, find row for SNP locus i, replace the entry with the new state\n",
    "            current_geneset.loc[i, 'alias_symbol'] = newstate    \n",
    "\n",
    "            ## record the current_geneset (Gibbs sample) in the round_ct column \n",
    "            ## using a for loop here with itertuples is ~30x slower \n",
    "            current_geneset['key'] = current_geneset['SNP'] + \",\" + current_geneset['alias_symbol']\n",
    "            final_gene_counts.loc[final_gene_counts['key'].isin(current_geneset['key']), 'round_ct'] += 1\n",
    "            \n",
    "            ## no garbage collection here because it increases time for a loop by ~7x\n",
    "        ## end the if statement for only Gibbs sampling on SNP loci with > 1 candidate gene \n",
    "    ## end for loop of iterating thru each SNP loci    \n",
    "\n",
    "    ## a round is finished so increment this for stopping criteria \n",
    "    final_rounds += 1\n",
    "    if ((final_rounds % 50) == 0):\n",
    "        print(\"Final (post-burn-in) sampling, finished full round thru SNP loci:\", final_rounds, \"\\n\")\n",
    "\n",
    "    ## add together the previous counts and new round_ct -> get new sampling counts\n",
    "    final_gene_counts['round_ct'] += final_gene_counts['counts'] \n",
    "\n",
    "    ## starting with the second round, calculate dif\n",
    "    ##     between sampling probabilities from previous round and current round\n",
    "    if final_rounds > 1:\n",
    "        ## sampling frequencies from previous round\n",
    "        final_gene_counts['freq_previous'] = final_gene_counts['counts'] / ((final_rounds - 1) * SNPs_per_round)\n",
    "        ## sampling frequencies from current round\n",
    "        final_gene_counts['freq_now'] = final_gene_counts['round_ct'] / (final_rounds * SNPs_per_round)\n",
    "        ## calculate sum of squared differences, then square-root this sum. \n",
    "        dif = math.sqrt(pd.Series.sum((final_gene_counts['freq_now'] - final_gene_counts['freq_previous']) ** 2))\n",
    "        print(\"dif is\", dif, \"\\n\")\n",
    "\n",
    "    ## replace old counts with current counts. \n",
    "    final_gene_counts['counts'] = final_gene_counts['round_ct']\n",
    "    ## drop freq columns if they're there\n",
    "    final_gene_counts = final_gene_counts[['SNP', 'alias_symbol', 'counts', 'key']]  \n",
    "\n",
    "## print the time it took to run\n",
    "t1 = time.time()\n",
    "final_time = t1 - t0\n",
    "final_time\n",
    "\n",
    "print(\"Number of final (post-burn-in) rounds of Gibbs sampling:\", final_rounds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output\n",
    "Will write files and display the final dataframes.    \n",
    "\n",
    "**CHANGES from original code:**  \n",
    "* Reproducing the *Nature Methods* paper's Supplementary Table 1 as output  \n",
    "\n",
    "**Parameters:**\n",
    "* path for result files directory   \n",
    "* path for original type of result file    \n",
    "* path for result file of the same format as Supplementary Table 1 from the paper   \n",
    "\n",
    "**Remember to change these last two parameters between runs if you don't want to rewrite your files from previous runs.**   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "## before running, check that these paths are what you want them to be. \n",
    "##     Especially when you don't want to rewrite files from previous runs! \n",
    "results_dir_path = Path.cwd().joinpath('iRIGS_results')\n",
    "original_output_path = Path.cwd().joinpath('iRIGS_results', 'SCZ_trial2_original.txt')\n",
    "supplement_table_path = Path.cwd().joinpath('iRIGS_results', 'SCZ_trial2_supple1.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing and recording the final sampling probabilities!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Processing and recording the final sampling probabilities!\\n\")\n",
    "\n",
    "## calculate final sampling probabilities \n",
    "final_gene_counts['posterior_probability'] = final_gene_counts['counts'] / (final_rounds * SNPs_per_round)\n",
    "\n",
    "## final dataframe\n",
    "final_gene_counts = final_gene_counts.filter(items = ['SNP', 'alias_symbol', 'posterior_probability'])\n",
    "final_gene_counts.rename(columns = {'alias_symbol':'candidate_gene_symbol'}, inplace = True)\n",
    "final_gene_counts.sort_values(by = ['SNP', 'posterior_probability'], ascending = [True, False], \\\n",
    "                              ignore_index = True, inplace = True)\n",
    "\n",
    "## create directory for result files\n",
    "if results_dir_path.exists() is False:\n",
    "    results_dir_path.mkdir(parents = True)\n",
    "    \n",
    "## create original result file\n",
    "final_gene_counts.to_csv(original_output_path, sep = '\\t', index = False)\n",
    "\n",
    "## Build table like in Supplementary Table 1 of paper\n",
    "## count number of candidate genes per SNP locus\n",
    "supple1 = final_gene_counts.copy().groupby('SNP').head(1)\n",
    "\n",
    "SNPs_per_loci = final_gene_counts.copy().groupby('SNP').size()  ## this is a series with SNP as index\n",
    "SNPs_per_loci.name = 'number_of_candidates'\n",
    "SNPs_per_loci = SNPs_per_loci.reset_index()\n",
    "\n",
    "supple1 = supple1.merge(SNPs_per_loci, on = 'SNP')\n",
    "\n",
    "## reorder to match supplementary table \n",
    "supple1 = supple1[['candidate_gene_symbol', 'posterior_probability', 'SNP', 'number_of_candidates']]\n",
    "\n",
    "## create supplementary table 1 result file\n",
    "supple1.to_csv(supplement_table_path, sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
